---
title: "Goseq_Analysis"
output: 
  html_document: 
    keep_md: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
#Goseq for Drap_Oases_Plant6 for IPS result
#all the scripts for running IPS is on interproscan.sh on ~/interproscan/interproscan-5.26-65.0 path

```{r}
setwd("~/Ferula_RNAseq_Analysis/Goseq_Analysis")
```

# 1) Command Line Pre-melt Steps

```{r}
# Combined tsv files from /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/interproscan/interproscan_result/Drap_Oases 
#cat Drap_Oases_Plant6.part-*.fasta.dir/Drap_Oases_Plant6.part-*.fasta.tsv >> Drap_Oases_Plant6_interproscan.tsv 
#cat Drap_Oases_Plant6_interproscan.tsv | cut -f1,12-50 > Drap_Oases_Plant6_All_ID.tsv
# cat Drap_Oases_Plant6_interproscan.tsv | awk '{print $1,$NF}' | sed '/GO:/!d' | sort | uniq > Drap_Oases_Plant6_GO_rough # Getting Transcript ID and GO terms
#cat Drap_Oases_Plant6_interproscan.tsv | awk '{print $1,$NF}' | sed '/GO:/!d' | sort | uniq | wc
#23484   46968 1312015
#cat Drap_Oases_Plant6_GO_rough | sed 's/|/ /g' > Drap_Oases_Plant6_GO_rough2
# cat Drap_Oases_Plant6_GO_rough | awk 'BEGIN{FS=" "}{print NF}' | sort | uniq -c # Finding lines with X fields (Max fields: 8)
#  10712 2
#    7358 3
#   3560 4
#   1586 5
#    235 6
#     32 7
#      1 8
# cat Drap_Oases_Plant6_GO_rough2 | sed ‘s/ /,/g’ > Drap_Oases_Plant6_GO_R # Changing field separator to comma
#nano Drap_Oases_Plant6_GO_R Added ,filler,filler,filler,filler,filler,filler # (so there are 8 columns/fields in line 1) Fixing column issues with R import

```

#2) Melting

```{r}
library(tidyverse)
library(reshape2)
Drap_Oases_Plant6_interpro_rough <- read.csv("~/interproscan/interproscan_result/Drap_Oases/Drap_Oases_Plant6_GO_R", ,header=FALSE, fill=TRUE)
dim(Drap_Oases_Plant6_interpro_rough)
#[1] 23484     9
Drap_Oases_Plant6_interpro_melt <- melt(Drap_Oases_Plant6_interpro_rough, id.vars = "V1", measure.vars=c("V2","V3","V4","V5","V6","V7","V8"))
write.table(Drap_Oases_Plant6_interpro_melt,file="/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6_interproscan_melt", quote=FALSE,row.names=FALSE)
```

# 3) Command Line Post Melt Formatting

```{r}
# cat Drap_Oases_Plant6_interproscan_melt | grep -v "V1 variable value" | awk '{print $1,$3}' | sort | uniq > Drap_Oases_Plant6_interproscan_edit # Removing leftover melt column
# cat Drap_Oases_Plant6_interproscan_edit | sed 's/ GO:/_GO:/g' | sed 's/ filler//g' | sed 's/ //g'> Drap_Oases_Plant6_interproscan_sep #Adding _ Separator for Removal of Extra ID Number
#cat Ae_interproscan_sep | awk -F_ '{$3="";print}' | sed 's/script /script_/g' | sed 's/  / /g' > Ae_interproscan_sep_edit # Removal of all Extra ID number
#cat Drap_Oases_Plant6_interproscan_sep |  sed "s/_GO/\tGO/" | grep GO | sed "s/_.\t\|_..\t\|_...\t/\t/" |  sed "s/_.\t\|_..\t\|_...\t/\t/" | sed "s/_.\t\|_..\t\|_...\t/\t/" > Drap_Oases_Plant6_interproscan_sep_edit #wc 41361
#The top script for my gene ID, becuase top do not work for me, with this remove the Extra ID number that added by running IPS (use this for downstream analysis)
#cat Drap_Oases_Plant6_interproscan_sep |  sed "s/_GO/\tGO/" | grep GO | sed "s/_.\t\|_..\t/\t/" > Drap_Oases_Plant6_interproscan_sep_edit #wc 41361
# cat Drap_Oases_Plant6_interproscan_sep_edit | sort | uniq > Drap_Oases_Plant6_interproscan_sep_edit_final #wc 39736 # Removing duplicate GOs for the same transcript
# cat Drap_Oases_Plant6_interproscan_sep_edit_final | grep "GO" > Drap_Oases_Plant6_GOseq #39736 Removing rows with no GO terms, it is same as previous, becuase I do remove no GO before this command
```

#For Transcript Files

```{r}
#cat Drap_Oases_Plant6_GOseq | awk '{print $1}' | sort | uniq > Drap_Oases_Plant6_ID # wc  19120 
```
#BnRNAseq for GO Enrichment (function)

```{r}
#1) GO annotaion 
library(ShortRead);library(goseq);library(GO.db);library("annotate");
#for ggplot heatmap library(heatmap)
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
Drap_Oases_Plant6.Bn.cdna<-readDNAStringSet("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6_No_Whitespace.fasta")
head(Drap_Oases_Plant6.Bn.cdna)
Drap_Oases_Plant6.bias<-nchar(Drap_Oases_Plant6.Bn.cdna) #60134
names(Drap_Oases_Plant6.bias)<-names(Drap_Oases_Plant6.Bn.cdna)
length(Drap_Oases_Plant6.bias) # 60134
# # bias.data vector must have the same length as DEgenes vector!
#convert to list (onetime procedure)
Drap_Oases_Plant6.Bngo<-read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6_GOseq",header=FALSE) # read Br_GO.txt in excel and save as csv
dim(Drap_Oases_Plant6.Bngo) #39736  2
head(head(Drap_Oases_Plant6.Bngo))
tail(head(Drap_Oases_Plant6.Bngo))
Drap_Oases_Plant6.Bngo.list <- tapply(as.character(Drap_Oases_Plant6.Bngo$V2),Drap_Oases_Plant6.Bngo$V1,c)
head(Drap_Oases_Plant6.Bngo.list)
save(Drap_Oases_Plant6.Bngo.list,file="/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.Bngo.list.Rdata") # this does not work ub goseq after updating R
# # Using manually entered categories.
# # Calculating the p-values...
# # 'select()' returned 1:1 mapping between keys and columns
Drap_Oases_Plant6.Bngo.DF<-as.data.frame(Drap_Oases_Plant6.Bngo.list)
dim(Drap_Oases_Plant6.Bngo.DF) #19120 1
Drap_Oases_Plant6.Bngo.DF$gene<-rownames(Drap_Oases_Plant6.Bngo.DF)
#Drap_Oases_Plant6.Bngo.DF[1:10,]
do.call(rbind.data.frame, Drap_Oases_Plant6.Bngo.list)
Drap_Oases_Plant6.Bngo.DF2<-do.call(rbind.data.frame,Drap_Oases_Plant6.Bngo.list) # ????

```

```{r}
load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.Bngo.list.Rdata")
#genelist could be any interested genes list (DEgenes list or all genes)
#first run for all genes
Drap_Oases_Plant6_ID<-read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6_ID",header=FALSE)
genelist <- Drap_Oases_Plant6_ID$V1 
#genelist2 <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\1", genelist) # just remove the last number after "_"
Drap_Oases_Plant6.GOseq.Bn.ORA <-function(genelist,padjust=0.05,ontology="BP") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt)
Drap_Oases_Plant6.DEgene.GO.ORA.gt$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt <- Drap_Oases_Plant6.DEgene.GO.ORA.gt[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt)
Drap_Oases_Plant6.gt.melt <- melt(Drap_Oases_Plant6.gt)
Drap_Oases_Plant6.gt.melt %>% length()
Drap_Oases_Plant6.gt.melt %>% head()
Drap_Oases_Plant6.gt.melt$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap1 <- ggplot(data = Drap_Oases_Plant6.gt.melt)
Drap_Oases_Plant6.heatmap1 <- Drap_Oases_Plant6.heatmap1 + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap1 <- Drap_Oases_Plant6.heatmap1 + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap1 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap1.png", width = 13, height = 25)

```
```{r}
load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.Bngo.list.Rdata")
#genelist could be any interested genes list (DEgenes list or all genes)
#first run for all genes
Drap_Oases_Plant6_ID<-read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6_ID",header=FALSE)
genelist <- Drap_Oases_Plant6_ID$V1 
#genelist2 <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\1", genelist) # just remove the last number after "_"
Drap_Oases_Plant6.GOseq.Bn.ORA.CC <-function(genelist,padjust=0.05,ontology="CC") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```

#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.CC <- Drap_Oases_Plant6.GOseq.Bn.ORA.CC(genelist)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.CC)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.CC$term
#play with pvalue
Drap_Oases_Plant6.DEgene.GO.ORA.gt.CC[Drap_Oases_Plant6.DEgene.GO.ORA.gt.CC$over_represented_pvalue < 0.05,]

write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.CC[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.CC", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.CC[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.CC.revigo", quote = FALSE,col.names = FALSE)
```

#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.CC <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.CC[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.CC)
Drap_Oases_Plant6.gt.melt.CC <- melt(Drap_Oases_Plant6.gt.CC)
Drap_Oases_Plant6.gt.melt.CC %>% length()
Drap_Oases_Plant6.gt.melt.CC %>% head()
Drap_Oases_Plant6.gt.melt.CC$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.CC$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap1.CC <- ggplot(data = Drap_Oases_Plant6.gt.melt.CC)
Drap_Oases_Plant6.heatmap1.CC <- Drap_Oases_Plant6.heatmap1.CC + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap1.CC <- Drap_Oases_Plant6.heatmap1.CC + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap1.CC 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap1.CC.png", width = 13, height = 25)
```


```{r}
load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.Bngo.list.Rdata")
#genelist could be any interested genes list (DEgenes list or all genes)
#first run for all genes
Drap_Oases_Plant6_ID<-read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6_ID",header=FALSE)
genelist <- Drap_Oases_Plant6_ID$V1 
#genelist2 <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\1", genelist) # just remove the last number after "_"
Drap_Oases_Plant6.GOseq.Bn.ORA.MF <-function(genelist,padjust=0.05,ontology="MF") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```

#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.MF <- Drap_Oases_Plant6.GOseq.Bn.ORA.MF(genelist)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.MF)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.MF$term
#play with pvalue
Drap_Oases_Plant6.DEgene.GO.ORA.gt.MF[Drap_Oases_Plant6.DEgene.GO.ORA.gt.MF$over_represented_pvalue < 0.05,]

write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.MF[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.MF", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.MF[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.MF.revigo", quote = FALSE,col.names = FALSE)
```

#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.MF <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.MF[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.MF)
Drap_Oases_Plant6.gt.melt.MF <- melt(Drap_Oases_Plant6.gt.MF)
Drap_Oases_Plant6.gt.melt.MF %>% length()
Drap_Oases_Plant6.gt.melt.MF %>% head()
Drap_Oases_Plant6.gt.melt.MF$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.MF$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap1.MF <- ggplot(data = Drap_Oases_Plant6.gt.melt.MF)
Drap_Oases_Plant6.heatmap1.MF <- Drap_Oases_Plant6.heatmap1.MF + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap1.MF <- Drap_Oases_Plant6.heatmap1.MF + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap1.MF 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap1.MF.png", width = 13, height = 25)








### test 
Drap_Oases_Plant6.GOseq.ID$V1 %>% head()
genes <- Drap_Oases_Plant6.GOseq.ID$V1
Drap_Oases_Plant6.Bngo[Drap_Oases_Plant6.Bngo$V2 == "GO:0043565",]
```

```{r}
expressed.genes <- read.delim("Goseq_Analysis/genelist",as.is=TRUE)
head(expressed.genes)
names(expressed.genes) <- "GeneID"
gene.lengths <- read.table("Goseq_Analysis/genes.length.txt2",as.is=TRUE)
head(gene.lengths)
summary(gene.lengths)
#we need to reduce the gene.length data to only contain entries for those genes in our expressed.genes set.  We also need this as a vector
gene.lengths.vector <- gene.lengths$V2[gene.lengths$V1 %in% expressed.genes$GeneID]
names(gene.lengths.vector) <- gene.lengths$V1[gene.lengths$V1 %in% expressed.genes$GeneID]
head(gene.lengths.vector)
expressed.genes.match <- expressed.genes[expressed.genes$GeneID %in% names(gene.lengths.vector),]
go.list <- strsplit(Drap_Oases_Plant6.BngoD$,split=",")
names(go.list) <- go.terms$GeneID
head(go.list)
nullp.result <- nullp(DEgenes = expressed.genes.match,bias.data = gene.lengths.vector)

```

#second for DEgenes 

```{r}
DEgenes2 <- topTags(dge.lrt.trt,n = Inf)$table[topTags(dge.lrt.trt,n = Inf)$table$FDR<0.05,]
write.csv(DEgenes2,file="~/Ferula_RNAseq_Analysis/Kallisto_out/Ferula_RNAseq_drap_oases_plant6_Davis_DEgenes2.csv")
genelist2 <- rownames(DEgenes2)
genelist2 %>% length() #3823
Drap_Oases_Plant6.GOseq.Bn.ORA2 <-function(genelist,padjust=0.05,ontology="BP") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plantx6.bias) %in% genelist2)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt2 <- Drap_Oases_Plant6.GOseq.Bn.ORA2(genelist2)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt2)
Drap_Oases_Plant6.DEgene.GO.ORA.gt2$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt2[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt2", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt2[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt2.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt2 <- Drap_Oases_Plant6.DEgene.GO.ORA.gt2[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt2)
Drap_Oases_Plant6.gt.melt2 <- melt(Drap_Oases_Plant6.gt2)
Drap_Oases_Plant6.gt.melt2 %>% length() #3
Drap_Oases_Plant6.gt.melt2 %>% head()
Drap_Oases_Plant6.gt.melt2$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt2$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap2 <- ggplot(data = Drap_Oases_Plant6.gt.melt2)
Drap_Oases_Plant6.heatmap2 <- Drap_Oases_Plant6.heatmap2 + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap2 <- Drap_Oases_Plant6.heatmap2 + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap2 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap2.png", width = 13, height = 25)
```

```{r}
#subset CC
genelist2 <- rownames(DEgenes2)
genelist2 %>% length() #3823
Drap_Oases_Plant6.GOseq.Bn.ORA2.CC <-function(genelist,padjust=0.05,ontology="CC") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist2)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt2.CC <- Drap_Oases_Plant6.GOseq.Bn.ORA2.CC(genelist2)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt2.CC)
Drap_Oases_Plant6.DEgene.GO.ORA.gt2.CC$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt2.CC[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt2.CC", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt2.CC[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt2.CC.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt2.CC <- Drap_Oases_Plant6.DEgene.GO.ORA.gt2.CC[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt2.CC)
Drap_Oases_Plant6.gt.melt2.CC <- melt(Drap_Oases_Plant6.gt2.CC)
Drap_Oases_Plant6.gt.melt2.CC %>% length() #3
Drap_Oases_Plant6.gt.melt2.CC %>% head()
Drap_Oases_Plant6.gt.melt2.CC$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt2.CC$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap2.CC <- ggplot(data = Drap_Oases_Plant6.gt.melt2.CC)
Drap_Oases_Plant6.heatmap2.CC <- Drap_Oases_Plant6.heatmap2.CC + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap2.CC <- Drap_Oases_Plant6.heatmap2.CC + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap2.CC 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap2.CC.png", width = 13, height = 25)
```

```{r}
#subset MF
genelist2 <- rownames(DEgenes2)
genelist2 %>% length() #3823
Drap_Oases_Plant6.GOseq.Bn.ORA2.MF <-function(genelist,padjust=0.05,ontology="MF") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist2)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt2.MF <- Drap_Oases_Plant6.GOseq.Bn.ORA2.MF(genelist2)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt2.MF)
Drap_Oases_Plant6.DEgene.GO.ORA.gt2.MF$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt2.MF[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt2.MF", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt2.MF[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt2.MF.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt2.MF <- Drap_Oases_Plant6.DEgene.GO.ORA.gt2.MF[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt2.MF)
Drap_Oases_Plant6.gt.melt2.MF <- melt(Drap_Oases_Plant6.gt2.MF)
Drap_Oases_Plant6.gt.melt2.MF %>% length() #3
Drap_Oases_Plant6.gt.melt2.MF %>% head()
Drap_Oases_Plant6.gt.melt2.MF$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt2.MF$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap2.MF <- ggplot(data = Drap_Oases_Plant6.gt.melt2.MF)
Drap_Oases_Plant6.heatmap2.MF <- Drap_Oases_Plant6.heatmap2.MF + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap2.MF <- Drap_Oases_Plant6.heatmap2.MF + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap2.MF 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap2.MF.png", width = 13, height = 25)
```


```{r}
#run for DEgenes but with the result of pairwise comparison 
#First FvsR
genelist.FvsR <- rownames(DEgene.FvsR)
genelist.FvsR %>% length() #578
Drap_Oases_Plant6.GOseq.Bn.ORA.FvsR <-function(genelist,padjust=0.05,ontology="BP") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist.FvsR)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR <- Drap_Oases_Plant6.GOseq.Bn.ORA.FvsR(genelist.FvsR)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.FvsR <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.FvsR)
Drap_Oases_Plant6.gt.melt.FvsR <- melt(Drap_Oases_Plant6.gt.FvsR)
Drap_Oases_Plant6.gt.melt.FvsR %>% length() #3
Drap_Oases_Plant6.gt.melt.FvsR %>% head()
Drap_Oases_Plant6.gt.melt.FvsR$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.FvsR$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap.FvsR <- ggplot(data = Drap_Oases_Plant6.gt.melt.FvsR)
Drap_Oases_Plant6.heatmap.FvsR <- Drap_Oases_Plant6.heatmap.FvsR + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap.FvsR <- Drap_Oases_Plant6.heatmap.FvsR + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap.FvsR 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap.FvsR.png", width = 13, height = 25)
```


```{r}
#run for DEgenes but with the result of pairwise comparison for subsetting CC
genelist.FvsR <- rownames(DEgene.FvsR)
genelist.FvsR %>% length() #578
Drap_Oases_Plant6.GOseq.Bn.ORA.FvsR.CC <-function(genelist,padjust=0.05,ontology="CC") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist.FvsR)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.CC <- Drap_Oases_Plant6.GOseq.Bn.ORA.FvsR.CC(genelist.FvsR)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.CC)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.CC$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.CC[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.CC", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.CC[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.CC.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.FvsR.CC <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.CC[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.FvsR.CC)
Drap_Oases_Plant6.gt.melt.FvsR.CC <- melt(Drap_Oases_Plant6.gt.FvsR.CC)
Drap_Oases_Plant6.gt.melt.FvsR.CC %>% length() #3
Drap_Oases_Plant6.gt.melt.FvsR.CC %>% head()
Drap_Oases_Plant6.gt.melt.FvsR.CC$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.FvsR.CC$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap.FvsR.CC <- ggplot(data = Drap_Oases_Plant6.gt.melt.FvsR.CC)
Drap_Oases_Plant6.heatmap.FvsR.CC <- Drap_Oases_Plant6.heatmap.FvsR.CC + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap.FvsR.CC <- Drap_Oases_Plant6.heatmap.FvsR.CC + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap.FvsR.CC 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap.FvsR.CC.png", width = 13, height = 25)
```

```{r}
#run for DEgenes but with the result of pairwise comparison for subsetting MF
genelist.FvsR <- rownames(DEgene.FvsR)
genelist.FvsR %>% length() #578
Drap_Oases_Plant6.GOseq.Bn.ORA.FvsR.MF <-function(genelist,padjust=0.05,ontology="MF") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist.FvsR)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.MF <- Drap_Oases_Plant6.GOseq.Bn.ORA.FvsR.MF(genelist.FvsR)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.MF)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.MF$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.MF[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.MF", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.MF[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.MF.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.FvsR.MF <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsR.MF[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.FvsR.MF)
Drap_Oases_Plant6.gt.melt.FvsR.MF <- melt(Drap_Oases_Plant6.gt.FvsR.MF)
Drap_Oases_Plant6.gt.melt.FvsR.MF %>% length() #3
Drap_Oases_Plant6.gt.melt.FvsR.MF %>% head()
Drap_Oases_Plant6.gt.melt.FvsR.MF$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.FvsR.MF$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap.FvsR.MF <- ggplot(data = Drap_Oases_Plant6.gt.melt.FvsR.MF)
Drap_Oases_Plant6.heatmap.FvsR.MF <- Drap_Oases_Plant6.heatmap.FvsR.MF + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap.FvsR.MF <- Drap_Oases_Plant6.heatmap.FvsR.MF + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap.FvsR.MF 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap.FvsR.MF.png", width = 13, height = 25)
```


```{r}
#Second SvsR
genelist.SvsR <- rownames(DEgene.SvsR)
genelist.SvsR %>% length() #5
Drap_Oases_Plant6.GOseq.Bn.ORA.SvsR <-function(genelist,padjust=0.05,ontology="BP") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist.SvsR)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.SvsR <- Drap_Oases_Plant6.GOseq.Bn.ORA.SvsR(genelist.SvsR) #no enriched GO
```

```{r}
#run for DEgenes but with the result of pairwise comparison for subsetting BP
#LvsR
genelist.LvsR <- rownames(DEgene.LvsR)
genelist.LvsR %>% length() #3779
Drap_Oases_Plant6.GOseq.Bn.ORA.LvsR <-function(genelist,padjust=0.05,ontology="BP") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist.LvsR)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR <- Drap_Oases_Plant6.GOseq.Bn.ORA.LvsR(genelist.LvsR)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.LvsR <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.LvsR)
Drap_Oases_Plant6.gt.melt.LvsR <- melt(Drap_Oases_Plant6.gt.LvsR)
Drap_Oases_Plant6.gt.melt.LvsR %>% length() #3
Drap_Oases_Plant6.gt.melt.LvsR %>% head()
Drap_Oases_Plant6.gt.melt.LvsR$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.LvsR$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap.LvsR <- ggplot(data = Drap_Oases_Plant6.gt.melt.LvsR)
Drap_Oases_Plant6.heatmap.LvsR <- Drap_Oases_Plant6.heatmap.LvsR + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap.LvsR <- Drap_Oases_Plant6.heatmap.LvsR + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap.LvsR 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap.LvsR.png", width = 13, height = 25)
```

```{r}
#run for DEgenes but with the result of pairwise comparison for subsetting CC
#LvsR
genelist.LvsR <- rownames(DEgene.LvsR)
genelist.LvsR %>% length() #3779
Drap_Oases_Plant6.GOseq.Bn.ORA.LvsR.CC <-function(genelist,padjust=0.05,ontology="CC") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist.LvsR)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC <- Drap_Oases_Plant6.GOseq.Bn.ORA.LvsR.CC(genelist.LvsR)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.LvsR.CC <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.LvsR.CC)
Drap_Oases_Plant6.gt.melt.LvsR.CC <- melt(Drap_Oases_Plant6.gt.LvsR.CC)
Drap_Oases_Plant6.gt.melt.LvsR.CC %>% length() #3
Drap_Oases_Plant6.gt.melt.LvsR.CC %>% head()
Drap_Oases_Plant6.gt.melt.LvsR.CC$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.LvsR.CC$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap.LvsR.CC <- ggplot(data = Drap_Oases_Plant6.gt.melt.LvsR.CC)
Drap_Oases_Plant6.heatmap.LvsR.CC <- Drap_Oases_Plant6.heatmap.LvsR.CC + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap.LvsR.CC <- Drap_Oases_Plant6.heatmap.LvsR.CC + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap.LvsR.CC 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap.LvsR.CC.png", width = 13, height = 25)
```

```{r}
#run for DEgenes but with the result of pairwise comparison for subsetting MF
#LvsR
genelist.LvsR <- rownames(DEgene.LvsR)
genelist.LvsR %>% length() #3779
Drap_Oases_Plant6.GOseq.Bn.ORA.LvsR.MF <-function(genelist,padjust=0.05,ontology="MF") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist.LvsR)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.MF <- Drap_Oases_Plant6.GOseq.Bn.ORA.LvsR.MF(genelist.LvsR)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.MF)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.MF$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.MF[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.MF", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.MF[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.MF.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.LvsR.MF <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.MF[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.LvsR.MF)
Drap_Oases_Plant6.gt.melt.LvsR.MF <- melt(Drap_Oases_Plant6.gt.LvsR.MF)
Drap_Oases_Plant6.gt.melt.LvsR.MF %>% length() #3
Drap_Oases_Plant6.gt.melt.LvsR.MF %>% head()
Drap_Oases_Plant6.gt.melt.LvsR.MF$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.LvsR.MF$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap.LvsR.MF <- ggplot(data = Drap_Oases_Plant6.gt.melt.LvsR.MF)
Drap_Oases_Plant6.heatmap.LvsR.MF <- Drap_Oases_Plant6.heatmap.LvsR.MF + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap.LvsR.MF <- Drap_Oases_Plant6.heatmap.LvsR.MF + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap.LvsR.MF 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap.LvsR.MF.png", width = 13, height = 25)
```

```{r}
```{r}
#run for DEgenes but with the result of pairwise comparison for subsetting CC
#LvsR
genelist.LvsR <- rownames(DEgene.LvsR)
genelist.LvsR %>% length() #3779
Drap_Oases_Plant6.GOseq.Bn.ORA.LvsR.CC <-function(genelist,padjust=0.05,ontology="CC") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist.LvsR)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC <- Drap_Oases_Plant6.GOseq.Bn.ORA.LvsR.CC(genelist.LvsR)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.LvsR.CC <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.LvsR.CC[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.LvsR.CC)
Drap_Oases_Plant6.gt.melt.LvsR.CC <- melt(Drap_Oases_Plant6.gt.LvsR.CC)
Drap_Oases_Plant6.gt.melt.LvsR.CC %>% length() #3
Drap_Oases_Plant6.gt.melt.LvsR.CC %>% head()
Drap_Oases_Plant6.gt.melt.LvsR.CC$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.LvsR.CC$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap.LvsR.CC <- ggplot(data = Drap_Oases_Plant6.gt.melt.LvsR.CC)
Drap_Oases_Plant6.heatmap.LvsR.CC <- Drap_Oases_Plant6.heatmap.LvsR.CC + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap.LvsR.CC <- Drap_Oases_Plant6.heatmap.LvsR.CC + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap.LvsR.CC 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap.LvsR.CC.png", width = 13, height = 25)
```

```{r}
#run for DEgenes but with the result of pairwise comparison for subsetting BP
#FvsS
genelist.FvsS <- rownames(DEgene.FvsS)
genelist.FvsS %>% length() #493
Drap_Oases_Plant6.GOseq.Bn.ORA.FvsS <-function(genelist,padjust=0.05,ontology="BP") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist.FvsS)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS <- Drap_Oases_Plant6.GOseq.Bn.ORA.FvsS(genelist.FvsS)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.FvsS <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.FvsS)
Drap_Oases_Plant6.gt.melt.FvsS <- melt(Drap_Oases_Plant6.gt.FvsS)
Drap_Oases_Plant6.gt.melt.FvsS %>% length() #3
Drap_Oases_Plant6.gt.melt.FvsS %>% head()
Drap_Oases_Plant6.gt.melt.FvsS$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.FvsS$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap.FvsS <- ggplot(data = Drap_Oases_Plant6.gt.melt.FvsS)
Drap_Oases_Plant6.heatmap.FvsS <- Drap_Oases_Plant6.heatmap.FvsS + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap.FvsS <- Drap_Oases_Plant6.heatmap.FvsS + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap.FvsS 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap.FvsS.png", width = 13, height = 25)
```

```{r}
```{r}
#run for DEgenes but with the result of pairwise comparison for subsetting BP
#FvsS
genelist.FvsS <- rownames(DEgene.FvsS)
genelist.FvsS %>% length() #493
Drap_Oases_Plant6.GOseq.Bn.ORA.FvsS.CC <-function(genelist,padjust=0.05,ontology="CC") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist.FvsS)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS.CC <- Drap_Oases_Plant6.GOseq.Bn.ORA.FvsS.CC(genelist.FvsS)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS.CC)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS.CC$term
#[1] "nucleus"
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.FvsS.CC <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS.CC[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.FvsS.CC)
Drap_Oases_Plant6.gt.melt.FvsS.CC <- melt(Drap_Oases_Plant6.gt.FvsS.CC)
Drap_Oases_Plant6.gt.melt.FvsS.CC %>% length() #3
Drap_Oases_Plant6.gt.melt.FvsS.CC %>% head()
Drap_Oases_Plant6.gt.melt.FvsS.CC$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.FvsS.CC$value)
```

```{r}
#run for DEgenes but with the result of pairwise comparison for subsetting MF
#FvsS
genelist.FvsS <- rownames(DEgene.FvsS)
genelist.FvsS %>% length() #493
Drap_Oases_Plant6.GOseq.Bn.ORA.FvsS.MF <-function(genelist,padjust=0.05,ontology="MF") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist.FvsS)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS.MF <- Drap_Oases_Plant6.GOseq.Bn.ORA.FvsS.MF(genelist.FvsS)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS.MF)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS.MF$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS.MF[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS.MF", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS.MF[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS.MF.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.FvsS.MF <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsS.MF[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.FvsS.MF)
Drap_Oases_Plant6.gt.melt.FvsS.MF <- melt(Drap_Oases_Plant6.gt.FvsS.MF)
Drap_Oases_Plant6.gt.melt.FvsS.MF %>% length() #3
Drap_Oases_Plant6.gt.melt.FvsS.MF %>% head()
 Drap_Oases_Plant6.gt.melt.FvsS.MF$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.FvsS.MF$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap.FvsS.MF <- ggplot(data = Drap_Oases_Plant6.gt.melt.FvsS.MF)
Drap_Oases_Plant6.heatmap.FvsS.MF <- Drap_Oases_Plant6.heatmap.FvsS.MF + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap.FvsS.MF <- Drap_Oases_Plant6.heatmap.FvsS.MF + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap.FvsS.MF 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap.FvsS.MF.png", width = 13, height = 25)
```

```{r}
#run for DEgenes but with the result of pairwise comparison for subsetting PB
#FvsL
genelist.FvsL <- rownames(DEgene.FvsL)
genelist.FvsL %>% length() #707
Drap_Oases_Plant6.GOseq.Bn.ORA.FvsL <-function(genelist,padjust=0.05,ontology="BP") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist.FvsL)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL <- Drap_Oases_Plant6.GOseq.Bn.ORA.FvsL(genelist.FvsL)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.FvsL <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.FvsL)
Drap_Oases_Plant6.gt.melt.FvsL <- melt(Drap_Oases_Plant6.gt.FvsL)
Drap_Oases_Plant6.gt.melt.FvsL %>% length() #3
Drap_Oases_Plant6.gt.melt.FvsL %>% head()
 Drap_Oases_Plant6.gt.melt.FvsL$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.FvsL$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap.FvsL <- ggplot(data = Drap_Oases_Plant6.gt.melt.FvsL)
Drap_Oases_Plant6.heatmap.FvsL <- Drap_Oases_Plant6.heatmap.FvsL + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap.FvsL <- Drap_Oases_Plant6.heatmap.FvsL + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap.FvsL 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap.FvsL.png", width = 13, height = 25)
```


```{r}
#run for DEgenes but with the result of pairwise comparison for subsetting CC
#FvsL
genelist.FvsL <- rownames(DEgene.FvsL)
genelist.FvsL %>% length() #707
Drap_Oases_Plant6.GOseq.Bn.ORA.FvsL.CC <-function(genelist,padjust=0.05,ontology="CC") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist.FvsL)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.CC <- Drap_Oases_Plant6.GOseq.Bn.ORA.FvsL.CC(genelist.FvsL)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.CC)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.CC$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.CC[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.CC", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.CC[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.CC.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.FvsL.CC <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.CC[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.FvsL.CC)
Drap_Oases_Plant6.gt.melt.FvsL.CC <- melt(Drap_Oases_Plant6.gt.FvsL.CC)
Drap_Oases_Plant6.gt.melt.FvsL.CC %>% length() #3
Drap_Oases_Plant6.gt.melt.FvsL.CC %>% head()
Drap_Oases_Plant6.gt.melt.FvsL.CC$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.FvsL.CC$value)
```

```{r}
#run for DEgenes but with the result of pairwise comparison for subsetting MF
#FvsL
genelist.FvsL <- rownames(DEgene.FvsL)
genelist.FvsL %>% length() #707
Drap_Oases_Plant6.GOseq.Bn.ORA.FvsL.MF <-function(genelist,padjust=0.05,ontology="MF") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist.FvsL)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```
#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.MF <- Drap_Oases_Plant6.GOseq.Bn.ORA.FvsL.MF(genelist.FvsL)
class(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.MF)
Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.MF$term
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.MF[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.MF", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.MF[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.MF.revigo", quote = FALSE,col.names = FALSE)
```
#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap for gt & gt:tissue effect genes 
Drap_Oases_Plant6.gt.FvsL.MF <- Drap_Oases_Plant6.DEgene.GO.ORA.gt.FvsL.MF[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt.FvsL.MF)
Drap_Oases_Plant6.gt.melt.FvsL.MF <- melt(Drap_Oases_Plant6.gt.FvsL.MF)
Drap_Oases_Plant6.gt.melt.FvsL.MF %>% length() #3
Drap_Oases_Plant6.gt.melt.FvsL.MF %>% head()
Drap_Oases_Plant6.gt.melt.FvsL.MF$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt.FvsL.MF$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap.FvsL.MF <- ggplot(data = Drap_Oases_Plant6.gt.melt.FvsL.MF)
Drap_Oases_Plant6.heatmap.FvsL.MF <- Drap_Oases_Plant6.heatmap.FvsL.MF + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap.FvsL.MF <- Drap_Oases_Plant6.heatmap.FvsL.MF + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap.FvsL.MF 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.heatmap.FvsL.MF.png", width = 13, height = 25)
```

```{r}
FvsR_LvsR.gt <- merge(Drap_Oases_Plant6.gt.FvsR, Drap_Oases_Plant6.gt.LvsR, by="Term", all=TRUE)
names(FvsR_LvsR.gt)[c(2:3)] <- c("FvsR", "LvsR")
FvsR_LvsR_FvsL.gt <- merge(FvsR_LvsR.gt,Drap_Oases_Plant6.gt.FvsL, by="Term",all=TRUE)
names(FvsR_LvsR_FvsL.gt)[4] <- "FvsL"
FvsR_LvsR_FvsL_FvsS.gt <- merge(FvsR_LvsR_FvsL.gt,Drap_Oases_Plant6.gt.FvsS, by="Term",all=TRUE)
names(FvsR_LvsR_FvsL_FvsS.gt)[5] <- "FvsS"
FvsR_LvsR_FvsL_FvsS.gt.melt <- melt(FvsR_LvsR_FvsL_FvsS.gt)
FvsR_LvsR_FvsL_FvsS.gt.melt 
FvsR_LvsR_FvsL_FvsS.gt.melt $logPvalue <- -log10(FvsR_LvsR_FvsL_FvsS.gt.melt $value)

# 2) Plot 
FvsR_LvsR_FvsL_FvsS.heatmap1 <- ggplot(data = FvsR_LvsR_FvsL_FvsS.gt.melt)
FvsR_LvsR_FvsL_FvsS.heatmap1 <- FvsR_LvsR_FvsL_FvsS.heatmap1 + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
FvsR_LvsR_FvsL_FvsS.heatmap1 <- FvsR_LvsR_FvsL_FvsS.heatmap1 + labs(y = "GO Term", x="Genotype", title="Ferula assafoetida Novel Genes GO Heatmap") 
FvsR_LvsR_FvsL_FvsS.heatmap1 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/FvsR_LvsR_FvsL_FvsS.heatmap1.png", width = 13, height = 25)
```


```{r}
FvsR_LvsR.gt.CC <- merge(Drap_Oases_Plant6.gt.FvsR.CC, Drap_Oases_Plant6.gt.LvsR.CC, by="Term", all=TRUE)
names(FvsR_LvsR.gt.CC)[c(2:3)] <- c("FvsR", "LvsR")
FvsR_LvsR_FvsL.gt.CC <- merge(FvsR_LvsR.gt.CC,Drap_Oases_Plant6.gt.FvsL.CC, by="Term",all=TRUE)
names(FvsR_LvsR_FvsL.gt.CC)[4] <- "FvsL"
FvsR_LvsR_FvsL_FvsS.gt.CC <- merge(FvsR_LvsR_FvsL.gt.CC,Drap_Oases_Plant6.gt.FvsS.CC, by="Term",all=TRUE)
names(FvsR_LvsR_FvsL_FvsS.gt.CC)[5] <- "FvsS"
FvsR_LvsR_FvsL_FvsS.gt.melt.CC <- melt(FvsR_LvsR_FvsL_FvsS.gt.CC)
FvsR_LvsR_FvsL_FvsS.gt.melt.CC 
FvsR_LvsR_FvsL_FvsS.gt.melt.CC$logPvalue <- -log10(FvsR_LvsR_FvsL_FvsS.gt.melt.CC$value)

# 2) Plot 
FvsR_LvsR_FvsL_FvsS.heatmap1.CC <- ggplot(data = FvsR_LvsR_FvsL_FvsS.gt.melt.CC)
FvsR_LvsR_FvsL_FvsS.heatmap1.CC <- FvsR_LvsR_FvsL_FvsS.heatmap1.CC + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
FvsR_LvsR_FvsL_FvsS.heatmap1.CC <- FvsR_LvsR_FvsL_FvsS.heatmap1.CC + labs(y = "GO Term", x="Genotype", title="Ferula assafoetida Novel Genes GO Heatmap") 
FvsR_LvsR_FvsL_FvsS.heatmap1.CC 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/FvsR_LvsR_FvsL_FvsS.heatmap1.CC.png", width = 13, height = 25)
```

```{r}
FvsR_LvsR.gt.MF <- merge(Drap_Oases_Plant6.gt.FvsR.MF, Drap_Oases_Plant6.gt.LvsR.MF, by="Term", all=TRUE)
names(FvsR_LvsR.gt.MF)[c(2:3)] <- c("FvsR", "LvsR")
FvsR_LvsR_FvsL.gt.MF <- merge(FvsR_LvsR.gt.MF,Drap_Oases_Plant6.gt.FvsL.MF, by="Term",all=TRUE)
names(FvsR_LvsR_FvsL.gt.MF)[4] <- "FvsL"
FvsR_LvsR_FvsL_FvsS.gt.MF <- merge(FvsR_LvsR_FvsL.gt.MF,Drap_Oases_Plant6.gt.FvsS.MF, by="Term",all=TRUE)
names(FvsR_LvsR_FvsL_FvsS.gt.MF)[5] <- "FvsS"
FvsR_LvsR_FvsL_FvsS.gt.melt.MF <- melt(FvsR_LvsR_FvsL_FvsS.gt.MF)
FvsR_LvsR_FvsL_FvsS.gt.melt.MF 
FvsR_LvsR_FvsL_FvsS.gt.melt.MF$logPvalue <- -log10(FvsR_LvsR_FvsL_FvsS.gt.melt.MF$value)

# 2) Plot 
FvsR_LvsR_FvsL_FvsS.heatmap1.MF <- ggplot(data = FvsR_LvsR_FvsL_FvsS.gt.melt.MF)
FvsR_LvsR_FvsL_FvsS.heatmap1.MF <- FvsR_LvsR_FvsL_FvsS.heatmap1.MF + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
FvsR_LvsR_FvsL_FvsS.heatmap1.MF <- FvsR_LvsR_FvsL_FvsS.heatmap1.MF + labs(y = "GO Term", x="Genotype", title="Ferula assafoetida Novel Genes GO Heatmap") 
FvsR_LvsR_FvsL_FvsS.heatmap1.MF 
ggsave("~/Ferula_RNAseq_Analysis/Goseq_Analysis/FvsR_LvsR_FvsL_FvsS.heatmap1.MF.png", width = 13, height = 25)
```

