---
title: "GOseq_b2g_analysis"
output: 
  html_document: 
    keep_md: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Goseq for Drap_Oases_Plant6 for b2g result with "03_01_18_try" folder 
#all the scripts for running b2g is on 03_01_18_try folder

```{r}
setwd("~/Ferula_RNAseq_Analysis/GOseq_b2g_analysis")
```

# 1) Command Line Pre-melt Steps

```{r}
# b2g_project.txt files from /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/03_01_18_try
#cat b2g_project.txt | grep GO | cut -f1,10 > b2g_project_GO_rough # Getting Transcript ID and GO terms
#63637 127276 10398003
#cat b2g_project_GO_rough2 | sed 's/;/ /g' | awk 'BEGIN{FS=" "}{print NF}'| sort -n | uniq -c# Finding lines with X fields (Max fields: 210)
#tail -63636 b2g_project_GO_rough > b2g_project_GO_rough2
#less b2g_project_GO_rough2 | sed 's/    /;/' | sed 's/;/,/g' > b2g_project_GO_R # Changing field separator to comma
# nano b2g_project_GO_R: Added #210 filler, (so there are 210 columns/fields in line 1) Fixing column issues with R import

```

#2) Melting

```{r}
library(tidyverse)
library(reshape2)
b2g_project_interpro_rough <- read.csv("~/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/b2g_project_GO_R", header=FALSE, fill=TRUE)
dim(b2g_project_interpro_rough)
#[1]63636  210
b2g_project_interpro_melt <- melt(b2g_project_interpro_rough, id.vars = "V1", measure.vars=c("V2","V3","V4","V5","V6","V7","V8","V9","V10","V11","V12","V13","V14","V15","V16","V17","V18","V19","V20","V21","V22","V23","V24","V25","V26","V27","V28","V29","V30","V31","V32","V33","V34","V35","V36","V37","V38","V39","V40","V41","V42","V43","V44","V45","V46","V47","V48","V49","V50","V51","V52","V53","V54","V55","V56","V57","V58","V59","V60","V61","V62","V63","V64","V65","V66","V67","V68","V69","V70","V71","V72","V73","V74","V75","V76","V77","V78","V79","V80","V81","V82","V83","V84","V85","V86","V87","V88","V89","V90","V91","V92","V93","V94","V95","V96","V97","V98","V99","V100","V101","V102","V103","V104","V105","V106","V107","V108","V109","V110","V111","V112","V113","V114","V115","V116","V117","V118","V119","V120","V121","V122","V123","V124","V125","V126","V127","V128","V129","V130","V131","V132","V133","V134","V135","V136","V137","V138","V139","V140","V141","V142","V143","V144","V145","V146","V147","V148","V149","V150","V151","V152","V153","V154","V155","V156","V157","V158","V159","V160","V161","V162","V163","V164","V165","V166","V167","V168","V169","V170","V171","V172","V173","V174","V175","V176","V177","V178","V179","V180","V181","V182","V183","V184","V185","V186","V187","V188","V189","V190","V191","V192","V193","V194","V195","V196","V197","V198","V199","V200","V201","V202","V203","V204","V205","V206","V207","V208","V209","V210"))
b2g_project_interpro_melt %>% dim() #13299924   3
write.table(b2g_project_interpro_melt,file="/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/b2g_project_interpro_melt", quote=FALSE,row.names=FALSE)
#save(b2g_project_interpro_melt, file = "~/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/b2g_project_interpro_melt.RData")
```
# 3) Command Line Post Melt Formatting

```{r}
#cat b2g_project_interpro_melt | grep -v "V1 variable value" | awk '{print $1,$3}'| sort | uniq >  b2g_project_edit
#cat b2g_project_edit | sed 's/ filler//g' | sed 's/_ORF//g' > b2g_project_edit2
#cat b2g_project_edit2 | sort | uniq | grep "GO" > b2g_project_edit3
#for attributing all GO term to each contig
#test <- read.delim("b2g_project_edit3", header = F, sep = "\t")
View(test)
test <- read.delim("b2g_project_edit3", header = F, sep = " ")
#library(reshape2)
#library(tidyverse)
m.test <- melt(test)
View(m.test)
#the helpfull function for combining second column based on first column
aaa <- test %>% group_by(V1) %>% arrange(V2) %>% summarise(V2 = paste(V2, collapse =";"))
View(aaa)
aaa %>% dim() #47134 2
b2g_project_GO_final <- aaa #need this file for GO enrichment
#save(b2g_project_GO_final, file = "~/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/b2g_project_edit_final.RData")
#write.csv(b2g_project_GO_final,"~/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/b2g_project_GO_final")
```

#For Transcript Files

```{r}
#cat b2g_project_GO_final | sed 's/"//g' | sed 's/,/ /g' | awk '{print $2,$3}' > b2g_project_GO_final2
#tail -47134 b2g_project_GO_final2 > b2g_project_GO_final3
#cat b2g_project_GO_final3 | awk '{print $1}' | sort | uniq > b2g_project_ID # wc 47134  47135 
```
#GO Enrichment (function)

```{r}
Drap_Oases_Plant6.Bngo<-read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/GOseq_b2g_input_out/b2g_project_GO_final3",header=FALSE, stringsAsFactors = F) #add stringsAsFactors to fix the prob of creation list in next step
#convert Drap_Oases_Plant6.Bngo to list for using in downstream analysis, it is very important to have the list for ech object
View(Drap_Oases_Plant6.Bngo)
aaa <- as.list(Drap_Oases_Plant6.Bngo$V2)
names(aaa) <- Drap_Oases_Plant6.Bngo$V1
head(aaa)
Drap_Oases_Plant6.Bngo.list <- aaa #use this list for downstreaming analysis
#1) GO annotaion 
library(ShortRead);library(goseq);library(GO.db);library("annotate");
#for ggplot heatmap library(heatmap)
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
Drap_Oases_Plant6.cdna<-readDNAStringSet("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Goseq_interpro_input_out/Drap_Oases_Plant6_No_Whitespace.fasta")
head(Drap_Oases_Plant6.cdna)
Drap_Oases_Plant6.bias<-nchar(Drap_Oases_Plant6.cdna) #60134
names(Drap_Oases_Plant6.bias)<-names(Drap_Oases_Plant6.cdna)
length(Drap_Oases_Plant6.bias) # 60134
#Drap_Oases_Plant6.Bngo.list <- aaa
head(Drap_Oases_Plant6.Bngo.list)
save(Drap_Oases_Plant6.Bngo.list,file="/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/Drap_Oases_Plant6.Bngo.list.Rdata")
Drap_Oases_Plant6.Bngo.DF<-as.data.frame(Drap_Oases_Plant6.Bngo.list)
dim(Drap_Oases_Plant6.Bngo.DF) #47134 1
Drap_Oases_Plant6.Bngo.DF$gene<-rownames(Drap_Oases_Plant6.Bngo.DF)
#Drap_Oases_Plant6.Bngo.DF[1:10,]
do.call(rbind.data.frame, Drap_Oases_Plant6.Bngo.list)
#Drap_Oases_Plant6.Bngo.DF2<-do.call(rbind.data.frame,Drap_Oases_Plant6.Bngo.list) # ????
```

```{r}
load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/Drap_Oases_Plant6.Bngo.list.Rdata")
#genelist could be any interested genes list (DEgenes list or all genes)
#first run for all genes
Drap_Oases_Plant6_ID<-read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/b2g_project_ID",header=FALSE)
#first for all genes
genelist <- Drap_Oases_Plant6_ID$V1 
length(genelist)
Drap_Oases_Plant6.GOseq.ORA <-function(genelist,padjust=0.05,ontology="BP") { #return GO enrichment table, padjus, padjust=0.05
#remove additional ID after running IPS
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  lengGO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.DF,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```

#BnRNAseq for GO Enrichment (run)

```{r}
library(goseq)
Drap_Oases_Plant6.GO.ORA.gt <- Drap_Oases_Plant6.GOseq.ORA(genelist)
class(Drap_Oases_Plant6.GO.ORA.gt)
Drap_Oases_Plant6.GO.ORA.gt$term
write.table(Drap_Oases_Plant6.GO.ORA.gt[,c(1,2,6)],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.DEgene.GO.ORA.gt[,1:2],row.names=FALSE,file="~/Ferula_RNAseq_Analysis/Goseq_Analysis/Drap_Oases_Plant6.DEgene.GO.ORA.gt.revigo", quote = FALSE,col.names = FALSE)
```
