---
title: "GOseq_b2g_analysis"
output: 
  html_document: 
    keep_md: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Goseq for Drap_Oases_Plant6 for b2g result with "03_01_18_try" folder 
#all the scripts for running b2g is on 03_01_18_try folder

```{r}
setwd("~/Ferula_RNAseq_Analysis/GOseq_b2g_analysis")
```

# 1) Command Line Pre-melt Steps

```{r}
# b2g_project.txt files from /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/03_01_18_try
#cat b2g_project.txt | grep GO | cut -f1,10 > b2g_project_GO_rough # Getting Transcript ID and GO terms
#63637 127276 10398003
#cat b2g_project_GO_rough2 | sed 's/;/ /g' | awk 'BEGIN{FS=" "}{print NF}'| sort -n | uniq -c# Finding lines with X fields (Max fields: 210)
#tail -63636 b2g_project_GO_rough > b2g_project_GO_rough2
#less b2g_project_GO_rough2 | sed 's/    /;/' | sed 's/;/,/g' > b2g_project_GO_R # Changing field separator to comma
# nano b2g_project_GO_R: Added #210 filler, (so there are 210 columns/fields in line 1) Fixing column issues with R import

```

#2) Melting

```{r}
library(tidyverse)
library(reshape2)
b2g_project_interpro_rough <- read.csv("~/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/b2g_project_GO_R", header=FALSE, fill=TRUE)
dim(b2g_project_interpro_rough)
#[1]63636  210
b2g_project_interpro_melt <- melt(b2g_project_interpro_rough, id.vars = "V1", measure.vars=c("V2","V3","V4","V5","V6","V7","V8","V9","V10","V11","V12","V13","V14","V15","V16","V17","V18","V19","V20","V21","V22","V23","V24","V25","V26","V27","V28","V29","V30","V31","V32","V33","V34","V35","V36","V37","V38","V39","V40","V41","V42","V43","V44","V45","V46","V47","V48","V49","V50","V51","V52","V53","V54","V55","V56","V57","V58","V59","V60","V61","V62","V63","V64","V65","V66","V67","V68","V69","V70","V71","V72","V73","V74","V75","V76","V77","V78","V79","V80","V81","V82","V83","V84","V85","V86","V87","V88","V89","V90","V91","V92","V93","V94","V95","V96","V97","V98","V99","V100","V101","V102","V103","V104","V105","V106","V107","V108","V109","V110","V111","V112","V113","V114","V115","V116","V117","V118","V119","V120","V121","V122","V123","V124","V125","V126","V127","V128","V129","V130","V131","V132","V133","V134","V135","V136","V137","V138","V139","V140","V141","V142","V143","V144","V145","V146","V147","V148","V149","V150","V151","V152","V153","V154","V155","V156","V157","V158","V159","V160","V161","V162","V163","V164","V165","V166","V167","V168","V169","V170","V171","V172","V173","V174","V175","V176","V177","V178","V179","V180","V181","V182","V183","V184","V185","V186","V187","V188","V189","V190","V191","V192","V193","V194","V195","V196","V197","V198","V199","V200","V201","V202","V203","V204","V205","V206","V207","V208","V209","V210"))
b2g_project_interpro_melt %>% dim() #13299924   3
write.table(b2g_project_interpro_melt,file="/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/b2g_project_interpro_melt", quote=FALSE,row.names=FALSE)
#save(b2g_project_interpro_melt, file = "~/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/b2g_project_interpro_melt.RData")
```
# 3) Command Line Post Melt Formatting

```{r}
#cat b2g_project_interpro_melt | grep -v "V1 variable value" | awk '{print $1,$3}'| sort | uniq >  b2g_project_edit
#cat b2g_project_edit | sed 's/ filler//g' | sed 's/_ORF//g' > b2g_project_edit2
#cat b2g_project_edit2 | sort | uniq | grep "GO" > b2g_project_edit3
#for attributing all GO term to each contig
#test <- read.delim("b2g_project_edit3", header = F, sep = "\t")
View(test)
test <- read.delim("b2g_project_edit3", header = F, sep = " ")
#test[test$V2 == "GO:0000786",] #check the specific GO term and find the genes ID for this GO term
#convert Drap_Oases_Plant6.Bngo to list for using in downstream analysis, it is very important to have the list for each object, This value use for GOseq function
test.list <- tapply(as.character(test$V2),test$V1,c) #use this value for GOseq function
#library(reshape2)
#library(tidyverse)
#m.test <- melt(test)
#View(m.test)
#the helpfull function for combining second column based on first column
#aaa <- test %>% group_by(V1) %>% arrange(V2) %>% summarise(V2 = paste(V2, collapse =";"))
#View(aaa)
#aaa %>% dim() #47134 2
#b2g_project_GO_final <- aaa # I do not use this value for downstream analysis
#save(b2g_project_GO_final, file = "~/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/b2g_project_edit_final.RData")
#write.csv(b2g_project_GO_final,"~/Ferula_RNAseq_Analysis/GOseq_b2g_analysis/b2g_project_GO_final")
```

#For Transcript Files

```{r}
#cat b2g_project_GO_final | sed 's/"//g' | sed 's/,/ /g' | awk '{print $2,$3}' > b2g_project_GO_final2
#tail -47134 b2g_project_GO_final2 > b2g_project_GO_final3
#cat b2g_project_GO_final3 | awk '{print $1}' | sort | uniq > b2g_project_ID # wc 47134  47135 
```
#GO Enrichment (function)

```{r}
Drap_Oases_Plant6.Bngo<-read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/GOseq_b2g_input_out/b2g_project_GO_final3",header=FALSE, stringsAsFactors = F) #add stringsAsFactors to fix the prob of creation list in next step
#convert Drap_Oases_Plant6.Bngo to list for using in downstream analysis, it is very important to have the list for ech object, but this did not work for GOseq function,we need to seperate each GO term by "" 
View(Drap_Oases_Plant6.Bngo)
#aaa <- as.list(Drap_Oases_Plant6.Bngo$V2)
#names(aaa) <- Drap_Oases_Plant6.Bngo$V1
#head(aaa)
Drap_Oases_Plant6.Bngo.list <- test.list #use this list for downstreaming analysis
#1) GO annotaion 
library(ShortRead);library(goseq);library(GO.db);library("annotate");
#for ggplot heatmap library(heatmap)
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
Drap_Oases_Plant6.cdna<-readDNAStringSet("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Goseq_interpro_input_out/Drap_Oases_Plant6_No_Whitespace.fasta")
head(Drap_Oases_Plant6.cdna)
Drap_Oases_Plant6.bias<-nchar(Drap_Oases_Plant6.cdna) #60134
names(Drap_Oases_Plant6.bias)<-names(Drap_Oases_Plant6.cdna)
length(Drap_Oases_Plant6.bias) # 60134
save(Drap_Oases_Plant6.Bngo.list,file="/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/GOseq_b2g_input_out/Drap_Oases_Plant6.Bngo.list.Rdata")
##Drap_Oases_Plant6.Bngo.DF<-as.data.frame(Drap_Oases_Plant6.Bngo.list)
#dim(Drap_Oases_Plant6.Bngo.DF) #47134 1
#Drap_Oases_Plant6.Bngo.DF$gene<-rownames(Drap_Oases_Plant6.Bngo.DF)
#Drap_Oases_Plant6.Bngo.DF[1:10,]
#do.call(rbind.data.frame, Drap_Oases_Plant6.Bngo.list)
#Drap_Oases_Plant6.Bngo.DF2<-do.call(rbind.data.frame,Drap_Oases_Plant6.Bngo.list) # ????
```

```{r}
#load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/GOseq_b2g_input_out/Drap_Oases_Plant6.Bngo.list.Rdata")
#genelist could be any interested genes list (DEgenes list or all genes)
#first run for all genes
Drap_Oases_Plant6_ID<-read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/GOseq_b2g_input_out/b2g_project_ID",header=FALSE)
#first for all genes
genelist <- Drap_Oases_Plant6_ID$V1 
length(genelist) #47134
Drap_Oases_Plant6.GOseq.Bn.ORA <-function(genelist,padjust=0.05,ontology="BP") { #return GO enrichment table, padjus, padjust=0.05
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```

#BnRNAseq for GO Enrichment (run)

```{r}
#library(goseq)
Drap_Oases_Plant6.GO.ORA.gt <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist)
class(Drap_Oases_Plant6.GO.ORA.gt)
Drap_Oases_Plant6.GO.ORA.gt$term
write.table(Drap_Oases_Plant6.GO.ORA.gt[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo", quote = FALSE,col.names = FALSE)
```

#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap 
Drap_Oases_Plant6.gt <- Drap_Oases_Plant6.GO.ORA.gt[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt)
Drap_Oases_Plant6.gt.melt <- melt(Drap_Oases_Plant6.gt)
Drap_Oases_Plant6.gt.melt %>% length()
Drap_Oases_Plant6.gt.melt %>% head()
Drap_Oases_Plant6.gt.melt$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap <- ggplot(data = Drap_Oases_Plant6.gt.melt)
Drap_Oases_Plant6.heatmap <- Drap_Oases_Plant6.heatmap + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap <- Drap_Oases_Plant6.heatmap + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap 
ggsave("~/GOseq_b2g_input_out/Drap_Oases_Plant6.heatmap.png", width = 13, height = 25) #it is unclear figure

```

#Subsetting MF

```{r}
#library(goseq)
Drap_Oases_Plant6.GO.ORA.gt.MF <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist,ontology = MF)
class(Drap_Oases_Plant6.GO.ORA.gt.MF)
Drap_Oases_Plant6.GO.ORA.gt.MF$term
write.table(Drap_Oases_Plant6.GO.ORA.gt.MF[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.MF", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.MF[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.MF", quote = FALSE,col.names = FALSE)
```
#Subsetting CC

```{r}
#library(goseq)
Drap_Oases_Plant6.GO.ORA.gt.CC <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist,ontology = "CC")
class(Drap_Oases_Plant6.GO.ORA.gt.CC)
Drap_Oases_Plant6.GO.ORA.gt.CC$term
write.table(Drap_Oases_Plant6.GO.ORA.gt.CC[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.CC", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.CC[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.CC", quote = FALSE,col.names = FALSE)
save(Drap_Oases_Plant6.GO.ORA.gt,Drap_Oases_Plant6.GO.ORA.gt.MF,Drap_Oases_Plant6.GO.ORA.gt.CC, file = "~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.term.Rdata")
```

```{r}
genelist <- Drap_Oases_Plant6_ID$V1 
length(genelist) #47134
Drap_Oases_Plant6.GOseq.Bn.ORA <-function(genelist,padjust=0.05,ontology="BP") { #return GO enrichment table, padjus, padjust=0.05
TF<-(names(Drap_Oases_Plant6.bias) %in% genelist)*1
names(TF)<-names(Drap_Oases_Plant6.bias)
pwf<-nullp(TF,bias.data=Drap_Oases_Plant6.bias)
  GO.pval <- goseq(pwf,gene2cat=Drap_Oases_Plant6.Bngo.list,use_genes_without_cat=TRUE) 
  if(ontology=="BP") {
    GO.pval2<-subset(GO.pval,ontology=="BP")
  } else if(ontology=="CC") {
    GO.pval2<-subset(GO.pval,ontology=="CC")
  } else {
    GO.pval2<-subset(GO.pval,ontology=="MF")
  }

GO.pval2$over_represented_padjust<-p.adjust(GO.pval2$over_represented_pvalue,method="BH")
  if(GO.pval2$over_represented_padjust[1]>padjust) stop("no enriched GO")
  else {
    enriched.GO<-GO.pval2[GO.pval2$over_represented_padjust<padjust,]
    print("enriched.GO is")
    print(enriched.GO)

    ## write Term and Definition
    for(i in 1:dim(enriched.GO)[1]) {
      enriched.GO$Term[i]<-Term(GOTERM[[enriched.GO[i,"category"]]])
      enriched.GO$Definition[i]<-Definition(GOTERM[[enriched.GO[i,"category"]]])
    }
    return(enriched.GO)
  }
}
```

```{r}
#library(goseq)
Drap_Oases_Plant6.GO.ORA.gt <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist)
class(Drap_Oases_Plant6.GO.ORA.gt)
Drap_Oases_Plant6.GO.ORA.gt$term
write.table(Drap_Oases_Plant6.GO.ORA.gt[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo", quote = FALSE,col.names = FALSE)
```

#GO Enrichment Heatmap (visualization)

```{r}
# 1) Loading Necessary Functions
library(WGCNA);library(ggplot2);library(reshape2);library(scales); library (plyr)
#draw heatmap 
Drap_Oases_Plant6.gt <- Drap_Oases_Plant6.GO.ORA.gt[,c("Term", "over_represented_padjust")] 
names(Drap_Oases_Plant6.gt)
Drap_Oases_Plant6.gt.melt <- melt(Drap_Oases_Plant6.gt)
Drap_Oases_Plant6.gt.melt %>% length()
Drap_Oases_Plant6.gt.melt %>% head()
Drap_Oases_Plant6.gt.melt$logPvalue <- -log10(Drap_Oases_Plant6.gt.melt$value)

# 2) Plot 
Drap_Oases_Plant6.heatmap <- ggplot(data = Drap_Oases_Plant6.gt.melt)
Drap_Oases_Plant6.heatmap <- Drap_Oases_Plant6.heatmap + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
Drap_Oases_Plant6.heatmap <- Drap_Oases_Plant6.heatmap + labs(y = "GO Term", title="Ferula assafoetida Novel Genes GO Heatmap") 
Drap_Oases_Plant6.heatmap 
ggsave("~/GOseq_b2g_input_out/Drap_Oases_Plant6.heatmap.png", width = 13, height = 25) #it is unclear figure
````

#pairwise comparison

```{r}
#Flower vs others
DEgene.FvsR <- read.csv("~/Kallisto_input_out/Ferula_RNAseq_drap_oases_plant6_Davis_DEgenes.FvsR.csv",header=T)
DEgene.FvsL <- read.csv("~/Kallisto_input_out/Ferula_RNAseq_drap_oases_plant6_Davis_DEgenes.FvsL.csv",header=T)
DEgene.FvsS <- read.csv("~/Kallisto_input_out/Ferula_RNAseq_drap_oases_plant6_Davis_DEgenes.FvsS.csv",header=T)
genelist <- DEgene.FvsR[,1] 
Drap_Oases_Plant6.GO.ORA.gt.FvsR <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist)
Drap_Oases_Plant6.gt.FvsR <- Drap_Oases_Plant6.GO.ORA.gt.FvsR[,c("Term", "over_represented_padjust")] 
Drap_Oases_Plant6.GO.ORA.gt.FvsR.MF <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist,ontology = "MF")
Drap_Oases_Plant6.gt.FvsR.MF <- Drap_Oases_Plant6.GO.ORA.gt.FvsR.MF[,c("Term", "over_represented_padjust")] 
Drap_Oases_Plant6.GO.ORA.gt.FvsR.CC <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist, ontology = "CC", padjust = 0.075) #no enriched GO

genelist <- DEgene.FvsL[,1] 
Drap_Oases_Plant6.GO.ORA.gt.FvsL <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist) #no enriched GO
Drap_Oases_Plant6.GO.ORA.gt.FvsL.MF <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist,ontology = "MF")
Drap_Oases_Plant6.gt.FvsL.MF <- Drap_Oases_Plant6.GO.ORA.gt.FvsL.MF[,c("Term", "over_represented_padjust")]
Drap_Oases_Plant6.GO.ORA.gt.FvsL.CC <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist,ontology = "CC")
Drap_Oases_Plant6.gt.FvsL.CC <- Drap_Oases_Plant6.GO.ORA.gt.FvsL.CC[,c("Term", "over_represented_padjust")]

genelist <- DEgene.FvsS[,1] 
Drap_Oases_Plant6.GO.ORA.gt.FvsS <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist) 
Drap_Oases_Plant6.gt.FvsS <- Drap_Oases_Plant6.GO.ORA.gt.FvsS[,c("Term", "over_represented_padjust")]

Drap_Oases_Plant6.GO.ORA.gt.FvsS.MF <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist,ontology = "MF")
Drap_Oases_Plant6.gt.FvsS.MF <- Drap_Oases_Plant6.GO.ORA.gt.FvsS.MF[,c("Term", "over_represented_padjust")]
Drap_Oases_Plant6.GO.ORA.gt.FvsS.CC <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist,ontology = "CC")
Drap_Oases_Plant6.gt.FvsS.CC <- Drap_Oases_Plant6.GO.ORA.gt.FvsS.CC[,c("Term", "over_represented_padjust")]

save(Drap_Oases_Plant6.GO.ORA.gt.FvsL.CC,Drap_Oases_Plant6.GO.ORA.gt.FvsL.MF,file = "~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.FvsL.term.Rdata")

save(Drap_Oases_Plant6.GO.ORA.gt.FvsS,Drap_Oases_Plant6.GO.ORA.gt.FvsS.MF,Drap_Oases_Plant6.GO.ORA.gt.FvsS.CC,file = "~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.FvsS.term.Rdata")

save(Drap_Oases_Plant6.GO.ORA.gt.FvsR,Drap_Oases_Plant6.GO.ORA.gt.FvsR.CC,Drap_Oases_Plant6.GO.ORA.gt.FvsR.MF,file = "~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.FvsR.term.Rdata")

```

```{r}
#Leaf vs others
DEgene.LvsR <- read.csv("~/Kallisto_input_out/Ferula_RNAseq_drap_oases_plant6_Davis_DEgenes.LvsR.csv",header=T)
genelist <- DEgene.LvsR[,1] 
Drap_Oases_Plant6.GO.ORA.gt.LvsR <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist)
Drap_Oases_Plant6.gt.LvsR <- Drap_Oases_Plant6.GO.ORA.gt.LvsR[,c("Term", "over_represented_padjust")]
Drap_Oases_Plant6.GO.ORA.gt.LvsR.MF <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist,ontology = "MF")
Drap_Oases_Plant6.gt.LvsR.MF <- Drap_Oases_Plant6.GO.ORA.gt.LvsR.MF[,c("Term", "over_represented_padjust")]

Drap_Oases_Plant6.GO.ORA.gt.LvsR.CC <- Drap_Oases_Plant6.GOseq.Bn.ORA(genelist, ontology = "CC") 
Drap_Oases_Plant6.gt.LvsR.CC <- Drap_Oases_Plant6.GO.ORA.gt.LvsR.CC[,c("Term", "over_represented_padjust")]

save(Drap_Oases_Plant6.GO.ORA.gt.LvsR,Drap_Oases_Plant6.GO.ORA.gt.LvsR.CC,Drap_Oases_Plant6.GO.ORA.gt.LvsR.MF,file = "~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.LvsR.term.Rdata")
```

#save

```{r}
write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsL.CC[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.FvsL.CC", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsL.CC[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.FvsL.CC", quote = FALSE,col.names = FALSE)

write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsL.MF[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.FvsL.MF", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsL.MF[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.FvsL.MF", quote = FALSE,col.names = FALSE)

write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsR.CC[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.FvsR.CC", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsR.CC[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.FvsR.CC", quote = FALSE,col.names = FALSE)

write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsR.MF[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.FvsR.MF", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsR.MF[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.FvsR.MF", quote = FALSE,col.names = FALSE)

write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsR[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.FvsR", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsR[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.FvsR", quote = FALSE,col.names = FALSE)

write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsS[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.FvsS", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsS[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.FvsS", quote = FALSE,col.names = FALSE)

write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsS.CC[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.FvsS.CC", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsS.CC[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.FvsS.CC", quote = FALSE,col.names = FALSE)

write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsS[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.FvsS", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsS[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.FvsS", quote = FALSE,col.names = FALSE)

write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsS.MF[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.FvsS.MF", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.FvsS.MF[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.FvsS.MF", quote = FALSE,col.names = FALSE)

write.table(Drap_Oases_Plant6.GO.ORA.gt.LvsR.CC[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.LvsR.CC", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.LvsR.CC[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.LvsR.CC", quote = FALSE,col.names = FALSE)

write.table(Drap_Oases_Plant6.GO.ORA.gt.LvsR.MF[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.LvsR.MF", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.LvsR.MF[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.LvsR.MF", quote = FALSE,col.names = FALSE)

write.table(Drap_Oases_Plant6.GO.ORA.gt.LvsR[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.LvsR", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.LvsR[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.LvsR", quote = FALSE,col.names = FALSE)
```

#Visualization

```{r}
#merge all pairwise comparision in to one figure

FvsR_LvsR.gt <- merge(Drap_Oases_Plant6.gt.FvsR, Drap_Oases_Plant6.gt.LvsR, by="Term", all=TRUE)
names(FvsR_LvsR.gt)[c(2:3)] <- c("FvsR", "LvsR")
FvsR_LvsR_FvsS.gt <- merge(FvsR_LvsR.gt,Drap_Oases_Plant6.gt.FvsS, by="Term",all=TRUE)
names(FvsR_LvsR_FvsS.gt)[4] <- "FvsS"
FvsR_LvsR_FvsS.gt.melt <- melt(FvsR_LvsR_FvsS.gt)
FvsR_LvsR_FvsS.gt.melt
FvsR_LvsR_FvsS.gt.melt$logPvalue <- -log10(FvsR_LvsR_FvsS.gt.melt$value)

# 2) Plot 
FvsR_LvsR_FvsS.heatmap <- ggplot(data = FvsR_LvsR_FvsS.gt.melt)
FvsR_LvsR_FvsS.heatmap <- FvsR_LvsR_FvsS.heatmap + geom_tile(color = "black", aes(x = factor(variable), y = Term, fill=logPvalue)) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 
FvsR_LvsR_FvsS.heatmap <- FvsR_LvsR_FvsS.heatmap + labs(y = "GO Term", x="Genotype", title="Ferula assafoetida DEgenes GO Heatmap") 
FvsR_LvsR_FvsS.heatmap 
ggsave("~/GOseq_b2g_input_out/FvsR_LvsR_FvsS.heatmap.png", width = 13, height = 25)
```

```{r}
genelist <- DEgene.FvsR[,1] 
Drap_Oases_Plant6.GO.ORA.gt.FvsR <- try(Drap_Oases_Plant6.GOseq.Bn.ORA(genelist))
Drap_Oases_Plant6.gt.FvsR <- Drap_Oases_Plant6.GO.ORA.gt.FvsR[,c("Term", "over_represented_padjust")] 
```

#run GO function for 44 modules,specify genelist for each of them

```{r}
files <- unlist(read.delim("~/WGCNA_input_out/Post_WGCNA_Analysis/file_list.txt", stringsAsFactors = F))
for (module in files){
  name <- paste0("~/WGCNA_input_out/Post_WGCNA_Analysis/", module)
  color <- sub("\\..*","", module)
  cat("Color is", color, "\n")
  module.genesID <- read.table(name, stringsAsFactors = F)
  genelist <- module.genesID$V2
  over_enriched <- try(Drap_Oases_Plant6.GOseq.Bn.ORA(genelist))
  if(class(over_enriched) != "try-error"){
    write.table(over_enriched[,c(1,2,6)],row.names=FALSE,file=paste0("~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.",color), quote = FALSE,col.names = TRUE)
    write.table(over_enriched[,1:2],row.names=FALSE,file=paste0("~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.revigo.", color), quote = FALSE,col.names = FALSE)
  }
}
violet.genesID <- read.table("~/WGCNA_input_out/Post_WGCNA_Analysis/violet.ModuleSelected.genesID_number.txt", stringsAsFactors = F)
genelist <- violet.genesID$V2 
Drap_Oases_Plant6.GO.ORA.gt.violet <- try(Drap_Oases_Plant6.GOseq.Bn.ORA(genelist)) #no enriched GO
Drap_Oases_Plant6.gt.violet<- Drap_Oases_Plant6.GO.ORA.gt.violet[,c("Term", "over_represented_padjust")]

write.table(Drap_Oases_Plant6.GO.ORA.gt.violet[,c(1,2,6)],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.violet", quote = FALSE,col.names = TRUE)
write.table(Drap_Oases_Plant6.GO.ORA.gt.violet[,1:2],row.names=FALSE,file="~/GOseq_b2g_input_out/Drap_Oases_Plant6.GO.ORA.gt.violet", quote = FALSE,col.names = FALSE)
#for loop????
```


