---
title: "KEGG_Analysis_GhostKOALA"
output: html_document
---
#copy and paste from jetstream server
#result are in /share/malooflab/Hajar/Annotation

#First get KO with using GhostKOALA and blastKOALA as Kazu's suggestion
# result save in ~/2017-ucdavis-igg201b/lab9/functional-analysis/functional-analysis_v2
```{r}
# reading GhoastKOALA results
library(tidyverse)
Ferula.KEGG<-read.csv(file.path("~/2017-ucdavis-igg201b/lab9/functional-analysis/functional-analysis_v2","Ferula.KEGG.GhostKOALA.txt"), col.names = FALSE) 
## Parsed with column specification:
## cols(
##   X1 = col_character()
#Needs to replace with BLASTKOALA results (annotation is on going as of April 12, 2018)
```

```{r}
head(Ferula.KEGG)
```

```{r}
Ferula.KEGG.s <- Ferula.KEGG %>% separate(FALSE.,c("gene","KO"),sep="\t")
head(Ferula.KEGG.s,100)
```

```{r}
Ferula.KEGG.s %>% is.na() %>% summary
 #    gene             KO         
 # Mode :logical   Mode :logical  
 # FALSE:43452     FALSE:16645  #near to Brassica (Kazu's result)  
 #                 TRUE :26807
```

```{r}
#convert into data.frame
Ferula.KEGG.s.DF <- as.data.frame(Ferula.KEGG.s)
dim(Ferula.KEGG.s.DF) #43452   2
save(Ferula.KEGG.s.DF,file = "~/2017-ucdavis-igg201b/lab9/functional-analysis/Ferula.KEGG.s.DF.RData")

```

```{r}
# KEGG enrichment analysis for leaf vs root
genes.LvsR <- read.csv("~/2017-ucdavis-igg201b/lab9/functional-analysis/Ferula_RNAseq_drap_oases_plant6_Davis_DEgenes.LvsR.csv", row.names = 1)
up.genes.LvsR <-subset(genes.LvsR, logFC > 1) #985 6
up.genes.names.LvsR <-row.names(up.genes.LvsR)
down.genes.LvsR <-subset(genes.LvsR, logFC < -1) #2794 6
down.genes.names.LvsR <-row.names(down.genes.LvsR) 
detach("package:dplyr")
# using enricher for general ORA purpose (https://guangchuangyu.github.io/tags/clusterprofiler/page/2/). No need to specify species like in enrichKEGG
library(clusterProfiler);library(KEGGREST)
#for install these libraries
# ~/R/x86_64-pc-linux-gnu-library
# sudo R , then run 
# source("https://bioconductor.org/biocLite.R")
# biocLite("DOSE")
```

```{r}
eKEGG.up.LvsR<-enricher(up.genes.names.LvsR,pvalueCutoff=0.05,pAdjustMethod="BH",TERM2GENE=Ferula.KEGG.s.DF[,c("KO","gene")])
eKEGG.down.LvsR<-enricher(down.genes.names.LvsR,pvalueCutoff=0.05,pAdjustMethod="BH",TERM2GENE=Ferula.KEGG.s.DF[,c("KO","gene")])
# convert "enrichResult" class into data frame
eKEGG.up.LvsR.DF<-as.data.frame(eKEGG.up.LvsR)
eKEGG.down.LvsR.DF<-as.data.frame(eKEGG.down.LvsR)
# add pathway id
eKEGG.up.LvsR.DF$DEF<-""
for(i in 1:nrow(eKEGG.up.LvsR.DF)) {
eKEGG.up.LvsR.DF$DEF[i]<-keggGet(as.character(eKEGG.up.LvsR.DF$ID[i]))[[1]]$DEFINITION
}

eKEGG.down.LvsR.DF$DEF<-""
for(i in 1:nrow(eKEGG.down.LvsR.DF)) {
eKEGG.down.LvsR.DF$DEF[i]<-keggGet(as.character(eKEGG.down.LvsR.DF$ID[i]))[[1]]$DEFINITION
}

```

```{r}
# add pathway id and name using KEGGREST package
eKEGG.up.LvsR.DF$pathway.name<-""
eKEGG.up.LvsR.DF$pathway.id<-""
for(i in 1:nrow(eKEGG.up.LvsR.DF)) {
    temp<-keggGet(as.character(eKEGG.up.LvsR.DF$ID[i]))[[1]]$PATHWAY
    print(temp)
  if(is.null(temp)) {
    eKEGG.up.LvsR.DF$pathway.name[i]<-"N/A"
    eKEGG.up.LvsR.DF$pathway.id[i]<-"N/A"
    } else {
    eKEGG.up.LvsR.DF$pathway.name[i]<-paste(temp,collapse="/")
    eKEGG.up.LvsR.DF$pathway.id[i]<-paste(names(temp),collapse="/")
  }
}

save(eKEGG.up.LvsR.DF, file =  "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.up.LvsR.DF.RData")

eKEGG.down.LvsR.DF$pathway.name<-""
eKEGG.down.LvsR.DF$pathway.id<-""
for(i in 1:nrow(eKEGG.down.LvsR.DF)) {
    temp<-keggGet(as.character(eKEGG.down.LvsR.DF$ID[i]))[[1]]$PATHWAY
    print(temp)
  if(is.null(temp)) {
    eKEGG.down.LvsR.DF$pathway.name[i]<-"N/A"
    eKEGG.down.LvsR.DF$pathway.id[i]<-"N/A"
    } else {
    eKEGG.down.LvsR.DF$pathway.name[i]<-paste(temp,collapse="/")
    eKEGG.down.LvsR.DF$pathway.id[i]<-paste(names(temp),collapse="/")
  }
}

save(eKEGG.down.LvsR.DF, file =  "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.down.LvsR.DF.RData")
```

```{r}
View(eKEGG.down.LvsR.DF)
# dotplot
pdf()
clusterProfiler::dotplot(eKEGG.down.LvsR)
dev.off()
```

```{r}
# ggplot geom_tile version
eKEGG.up.LvsR.DF$sample<-"Leaf vs Root"
eKEGG.up.LvsR.DF$FC<-"up"
eKEGG.down.LvsR.DF$sample<-"Leaf vs Root"
eKEGG.down.LvsR.DF$FC<-"down"
library(scales)
ggplot(eKEGG.up.LvsR.DF,aes(x=sample,y=DEF)) + geom_tile(aes(fill=-log10(p.adjust))) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) # OK

ggplot(eKEGG.down.LvsR.DF,aes(x=sample,y=DEF)) + geom_tile(aes(fill=-log10(p.adjust))) + scale_fill_gradient2(low=muted("green"), high=muted("magenta"))
```

```{r}
genes.FvsR <- read.csv("~/2017-ucdavis-igg201b/lab9/functional-analysis/Ferula_RNAseq_drap_oases_plant6_Davis_DEgenes.FvsR.csv", row.names = 1) #578
up.genes.FvsR <-subset(genes.LvsR, logFC > 1) #425 6
up.genes.names.FvsR <-row.names(up.genes.FvsR)
down.genes.FvsR <-subset(genes.FvsR, logFC < -1) #2794 6
down.genes.names.FvsR <-row.names(down.genes.FvsR) #153
detach("package:dplyr")
# using enricher for general ORA purpose (https://guangchuangyu.github.io/tags/clusterprofiler/page/2/). No need to specify species like in enrichKEGG
library(clusterProfiler);library(KEGGREST)
```

```{r}
eKEGG.up.FvsR<-enricher(up.genes.names.FvsR,pvalueCutoff=0.05,pAdjustMethod="BH",TERM2GENE=Ferula.KEGG.s.DF[,c("KO","gene")])
eKEGG.down.FvsR<-enricher(down.genes.names.FvsR,pvalueCutoff=0.05,pAdjustMethod="BH",TERM2GENE=Ferula.KEGG.s.DF[,c("KO","gene")])
# convert "enrichResult" class into data frame
eKEGG.up.FvsR.DF<-as.data.frame(eKEGG.up.FvsR)
eKEGG.down.FvsR.DF<-as.data.frame(eKEGG.down.FvsR)
# add pathway id
eKEGG.up.FvsR.DF$DEF<-""
for(i in 1:nrow(eKEGG.up.FvsR.DF)) {
eKEGG.up.FvsR.DF$DEF[i]<-keggGet(as.character(eKEGG.up.FvsR.DF$ID[i]))[[1]]$DEFINITION
}

eKEGG.down.FvsR.DF$DEF<-""
for(i in 1:nrow(eKEGG.down.FvsR.DF)) {
eKEGG.down.FvsR.DF$DEF[i]<-keggGet(as.character(eKEGG.down.FvsR.DF$ID[i]))[[1]]$DEFINITION
}

```

```{r}
# add pathway id and name using KEGGREST package
eKEGG.up.FvsR.DF$pathway.name<-""
eKEGG.up.FvsR.DF$pathway.id<-""
for(i in 1:nrow(eKEGG.up.FvsR.DF)) {
    temp<-keggGet(as.character(eKEGG.up.FvsR.DF$ID[i]))[[1]]$PATHWAY
    print(temp)
  if(is.null(temp)) {
    eKEGG.up.FvsR.DF$pathway.name[i]<-"N/A"
    eKEGG.up.FvsR.DF$pathway.id[i]<-"N/A"
    } else {
    eKEGG.up.FvsR.DF$pathway.name[i]<-paste(temp,collapse="/")
    eKEGG.up.FvsR.DF$pathway.id[i]<-paste(names(temp),collapse="/")
  }
}

save(eKEGG.up.FvsR.DF, file =  "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.up.FvsR.DF.RData")

eKEGG.down.FvsR.DF$pathway.name<-""
eKEGG.down.FvsR.DF$pathway.id<-""
for(i in 1:nrow(eKEGG.down.FvsR.DF)) {
    temp<-keggGet(as.character(eKEGG.down.FvsR.DF$ID[i]))[[1]]$PATHWAY
    print(temp)
  if(is.null(temp)) {
    eKEGG.down.FvsR.DF$pathway.name[i]<-"N/A"
    eKEGG.down.FvsR.DF$pathway.id[i]<-"N/A"
    } else {
    eKEGG.down.FvsR.DF$pathway.name[i]<-paste(temp,collapse="/")
    eKEGG.down.FvsR.DF$pathway.id[i]<-paste(names(temp),collapse="/")
  }
}

save(eKEGG.down.FvsR.DF, file =  "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.down.FvsR.DF.RData")
```

```{r}
View(eKEGG.down.FvsR.DF)
# dotplot
pdf()
clusterProfiler::dotplot(eKEGG.down.FvsR)
dev.off()

pdf()
clusterProfiler::dotplot(eKEGG.up.FvsR)
dev.off()
```

```{r}
# KEGG enrichment analysis for flower vs stem
genes.FvsS <- read.csv("~/2017-ucdavis-igg201b/lab9/functional-analysis/Ferula_RNAseq_drap_oases_plant6_Davis_DEgenes.FvsS.csv", row.names = 1) #493
up.genes.FvsS <-subset(genes.FvsS, logFC > 1) #413 5
up.genes.names.FvsS <-row.names(up.genes.FvsS) 
down.genes.FvsS <-subset(genes.FvsS, logFC < -1) #80 5
down.genes.names.FvsS <-row.names(down.genes.FvsS) 
detach("package:dplyr")
# using enricher for general ORA purpose (https://guangchuangyu.github.io/tags/clusterprofiler/page/2/). No need to specify species like in enrichKEGG
library(clusterProfiler);library(KEGGREST)
#for install these libraries
# ~/R/x86_64-pc-linux-gnu-library
# sudo R , then run 
# source("https://bioconductor.org/biocLite.R")
# biocLite("DOSE")
```

```{r}
load("~/2017-ucdavis-igg201b/lab9/functional-analysis/Ferula.KEGG.s.DF.RData")
eKEGG.up.FvsS<-enricher(up.genes.names.FvsS,pvalueCutoff=0.05,pAdjustMethod="BH",TERM2GENE=Ferula.KEGG.s.DF[,c("KO","gene")])
eKEGG.down.FvsS<-enricher(down.genes.names.FvsS,pvalueCutoff=0.05,pAdjustMethod="BH",TERM2GENE=Ferula.KEGG.s.DF[,c("KO","gene")])
# convert "enrichResult" class into data frame
eKEGG.up.FvsS.DF<-as.data.frame(eKEGG.up.FvsS)
eKEGG.down.FvsS.DF<-as.data.frame(eKEGG.down.FvsS)
# add pathway id
eKEGG.up.FvsS.DF$DEF<-""
for(i in 1:nrow(eKEGG.up.FvsS.DF)) {
eKEGG.up.FvsS.DF$DEF[i]<-keggGet(as.character(eKEGG.up.FvsS.DF$ID[i]))[[1]]$DEFINITION
}

eKEGG.down.FvsS.DF$DEF<-""
for(i in 1:nrow(eKEGG.down.FvsS.DF)) {
eKEGG.down.FvsS.DF$DEF[i]<-keggGet(as.character(eKEGG.down.FvsS.DF$ID[i]))[[1]]$DEFINITION
}

```

```{r}
# add pathway id and name using KEGGREST package
eKEGG.up.FvsS.DF$pathway.name<-""
eKEGG.up.FvsS.DF$pathway.id<-""
for(i in 1:nrow(eKEGG.up.FvsS.DF)) {
    temp<-keggGet(as.character(eKEGG.up.FvsS.DF$ID[i]))[[1]]$PATHWAY
    print(temp)
  if(is.null(temp)) {
    eKEGG.up.FvsS.DF$pathway.name[i]<-"N/A"
    eKEGG.up.FvsS.DF$pathway.id[i]<-"N/A"
    } else {
    eKEGG.up.FvsS.DF$pathway.name[i]<-paste(temp,collapse="/")
    eKEGG.up.FvsS.DF$pathway.id[i]<-paste(names(temp),collapse="/")
  }
}

save(eKEGG.up.FvsS.DF, file =  "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.up.FvsS.DF.RData")

eKEGG.down.FvsS.DF$pathway.name<-""
eKEGG.down.FvsS.DF$pathway.id<-""
for(i in 1:nrow(eKEGG.down.FvsS.DF)) {
    temp<-keggGet(as.character(eKEGG.down.FvsS.DF$ID[i]))[[1]]$PATHWAY
    print(temp)
  if(is.null(temp)) {
    eKEGG.down.FvsS.DF$pathway.name[i]<-"N/A"
    eKEGG.down.FvsS.DF$pathway.id[i]<-"N/A"
    } else {
    eKEGG.down.FvsS.DF$pathway.name[i]<-paste(temp,collapse="/")
    eKEGG.down.FvsS.DF$pathway.id[i]<-paste(names(temp),collapse="/")
  }
}

save(eKEGG.down.FvsS.DF, file =  "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.down.FvsS.DF.RData")
```


```{r}
# KEGG enrichment analysis for flower vs stem
genes.FvsL <- read.csv("~/2017-ucdavis-igg201b/lab9/functional-analysis/Ferula_RNAseq_drap_oases_plant6_Davis_DEgenes.FvsL.csv", row.names = 1) #707
up.genes.FvsL <-subset(genes.FvsL, logFC > 1) #642 5
up.genes.names.FvsL <-row.names(up.genes.FvsL) 
down.genes.FvsL <-subset(genes.FvsL, logFC < -1) #65 5
down.genes.names.FvsL <-row.names(down.genes.FvsL) 
detach("package:dplyr")
# using enricher for general ORA purpose (https://guangchuangyu.github.io/tags/clusterprofiler/page/2/). No need to specify species like in enrichKEGG
library(clusterProfiler);library(KEGGREST)
#for install these libraries
# ~/R/x86_64-pc-linux-gnu-library
# sudo R , then run 
# source("https://bioconductor.org/biocLite.R")
# biocLite("DOSE")
```

```{r}
#load("~/2017-ucdavis-igg201b/lab9/functional-analysis/Ferula.KEGG.s.DF.RData")
eKEGG.up.FvsL<-enricher(up.genes.names.FvsL,pvalueCutoff=0.05,pAdjustMethod="BH",TERM2GENE=Ferula.KEGG.s.DF[,c("KO","gene")])
eKEGG.down.FvsL<-enricher(down.genes.names.FvsL,pvalueCutoff=0.05,pAdjustMethod="BH",TERM2GENE=Ferula.KEGG.s.DF[,c("KO","gene")])
# convert "enrichResult" class into data frame
eKEGG.up.FvsL.DF<-as.data.frame(eKEGG.up.FvsL)
eKEGG.down.FvsL.DF<-as.data.frame(eKEGG.down.FvsL)
# add pathway id
eKEGG.up.FvsL.DF$DEF<-""
for(i in 1:nrow(eKEGG.up.FvsL.DF)) {
eKEGG.up.FvsL.DF$DEF[i]<-keggGet(as.character(eKEGG.up.FvsL.DF$ID[i]))[[1]]$DEFINITION
}

eKEGG.down.FvsL.DF$DEF<-""
for(i in 1:nrow(eKEGG.down.FvsL.DF)) {
eKEGG.down.FvsL.DF$DEF[i]<-keggGet(as.character(eKEGG.down.FvsL.DF$ID[i]))[[1]]$DEFINITION
}

```

```{r}
# add pathway id and name using KEGGREST package
eKEGG.up.FvsL.DF$pathway.name<-""
eKEGG.up.FvsL.DF$pathway.id<-""
for(i in 1:nrow(eKEGG.up.FvsL.DF)) {
    temp<-keggGet(as.character(eKEGG.up.FvsL.DF$ID[i]))[[1]]$PATHWAY
    print(temp)
  if(is.null(temp)) {
    eKEGG.up.FvsL.DF$pathway.name[i]<-"N/A"
    eKEGG.up.FvsL.DF$pathway.id[i]<-"N/A"
    } else {
    eKEGG.up.FvsL.DF$pathway.name[i]<-paste(temp,collapse="/")
    eKEGG.up.FvsL.DF$pathway.id[i]<-paste(names(temp),collapse="/")
  }
}

save(eKEGG.up.FvsL.DF, file =  "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.up.FvsL.DF.RData")

eKEGG.down.FvsL.DF$pathway.name<-""
eKEGG.down.FvsL.DF$pathway.id<-""
for(i in 1:nrow(eKEGG.down.FvsL.DF)) {
    temp<-keggGet(as.character(eKEGG.down.FvsL.DF$ID[i]))[[1]]$PATHWAY
    print(temp)
  if(is.null(temp)) {
    eKEGG.down.FvsL.DF$pathway.name[i]<-"N/A"
    eKEGG.down.FvsL.DF$pathway.id[i]<-"N/A"
    } else {
    eKEGG.down.FvsL.DF$pathway.name[i]<-paste(temp,collapse="/")
    eKEGG.down.FvsL.DF$pathway.id[i]<-paste(names(temp),collapse="/")
  }
}

save(eKEGG.down.FvsL.DF, file =  "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.down.FvsL.DF.RData")
```


```{r}
#put all of them in one figure
View(eKEGG.up.FvsS.DF)
# dotplot
load( "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.down.FvsS.DF.RData")
load( "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.up.FvsS.DF.RData")
load( "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.up.FvsL.DF.RData")
load( "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.down.FvsL.DF.RData")
load( "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.up.FvsR.DF.RData")
load( "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.down.FvsR.DF.RData")
load( "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.up.LvsR.DF.RData")
load( "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGG.down.LvsR.DF.RData")

#     
pdf(file = "plots6.pdf")
par(mfrow = c(1,2)) # not work try to put in one figure
clusterProfiler::dotplot(eKEGG.up.FvsS)
clusterProfiler::dotplot(eKEGG.down.FvsS)
dev.off()

#merge up and down for each comparison in one figure
#FvsS
a <- cbind(eKEGG.up.FvsS.DF, FC = "up")
b <- cbind(eKEGG.down.FvsS.DF, FC = "down")
c <- rbind(a,b)
pdf()
ggplot(c, aes(GeneRatio, ID)) +
  geom_point(aes(size = Count, color = p.adjust, shape = FC))
dev.off()

#FvsL
a <- cbind(eKEGG.up.FvsL.DF, FC = "up")
b <- cbind(eKEGG.down.FvsL.DF, FC = "down")
c <- rbind(a,b)
pdf()
ggplot(c, aes(GeneRatio, ID)) +
  geom_point(aes(size = Count, color = p.adjust, shape = FC))
dev.off()

#FvsR
a <- cbind(eKEGG.up.FvsR.DF, FC = "up")
b <- cbind(eKEGG.down.FvsR.DF, FC = "down")
c <- rbind(a,b)
pdf()
ggplot(c, aes(GeneRatio, ID)) +
  geom_point(aes(size = Count, color = p.adjust, shape = FC))
dev.off()

#LvsR
a <- cbind(eKEGG.up.LvsR.DF, FC = "up")
b <- cbind(eKEGG.down.LvsR.DF, FC = "down")
c <- rbind(a,b)
pdf()
ggplot(c, aes(GeneRatio, ID)) +
  geom_point(aes(size = Count, color = p.adjust, shape = FC))
dev.off()
```

```{r}
library(scales)
ggplot(eKEGG.up.FvsR.DF,aes(x=sample,y=DEF)) + geom_tile(aes(fill=-log10(p.adjust))) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) # OK
eKEGG.DF.up.down 

#visualize KEGG for one sample
ggplot(eKEGG.DF.up.down ,aes(x=sample,y=DEF)) + geom_tile(aes(fill=-log10(p.adjust))) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) 

Flower vs Root,
#summary of KEGG for all pairwise comparison
eKEGG.down.FvsR.DF$name <- "Flower vs Root"
eKEGG.up.FvsR.DF$name <- "Flower vs Root"
eKEGG.down.LvsR.DF$name <- "Leaf vs Root"
eKEGG.up.LvsR.DF$name <- "Leaf vs Root"
eKEGG.down.FvsL.DF$name <- "Flower vs Leaf"
eKEGG.up.FvsL.DF$name <- "Flower vs Leaf"
eKEGG.down.FvsS.DF$name <- "Flower vs Stem"
eKEGG.up.FvsS.DF$name <- "Flower vs Stem"

eKEGG.down.FvsR.DF$FC <- "down"
eKEGG.up.FvsR.DF$FC <- "up"
eKEGG.down.LvsR.DF$FC <- "down"
eKEGG.up.LvsR.DF$FC <- "up"
eKEGG.down.FvsS.DF$FC <- "down"
eKEGG.up.FvsS.DF$FC <- "up"
eKEGG.down.FvsL.DF$FC <- "down"
eKEGG.up.FvsL.DF$FC <- "up"

all <- rbind(eKEGG.up.FvsR.DF,eKEGG.up.LvsR.DF,eKEGG.down.FvsR.DF,eKEGG.down.LvsR.DF,eKEGG.up.FvsL.DF,eKEGG.down.FvsL.DF,eKEGG.up.FvsS.DF,eKEGG.down.FvsS.DF )

#visualize KEGG
p <- ggplot(all ,aes(x=name,y=DEF)) + geom_tile(aes(fill=-log10(p.adjust))) + scale_fill_gradient2(low=muted("green"), high=muted("magenta")) + facet_grid(~FC)
ggsave(p,filename = "~/2017-ucdavis-igg201b/lab9/functional-analysis/eKEGGS.plot.png", width = 15, height = 10)

```

#KEGG pathway map using pathview package

```{r}
library(pathview)
```

```{r}
map.ko00909 <- pathview::pathview(gene.data  = data.frame(row.names=eKEGG.up.FvsS.DF$ID[5],value=3), # KEGG ortholog IDs are also treated as gene IDs as to handle data. 
                     pathway.id = unlist(strsplit(eKEGG.up.FvsS.DF$pathway.id[5],split="/"))[1], 
                     species    = "ko"# When KEGG ortholog pathway is considered, species="ko". 
                     ) # 

map.ko04016 <- pathview::pathview(gene.data  = data.frame(row.names=all$ID[10],value=3), # KEGG ortholog IDs are also treated as gene IDs as to handle metagenomic data. 
                     pathway.id = unlist(strsplit(all$pathway.id[10],split="/"))[1], 
                     species    = "ko"
                     )# When KEGG ortholog pathway is considered, species="ko".

#result is same as Kazu????

```

