---
title: "Ferula_RNAseq_Kallisto_DEGs"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#input for this script after removing low counts of kallisto, I am thinking about using normalized read counts after removing batch effect (Ruijuan's scrip https://github.com/leejimmy93/KIAT/blob/master/F2/F2_network.Rmd) but regarding Limma's author "no need to use removeBatchEffects() when doing a differential expression analysis". 

```{r}
counts_All <- read.csv("/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/DEgenes_Analysis/Ferula_RNAseq_combined_drap_oases_plant6_counts_WGCNA_kalliso.csv", row.names = 1)
dim(counts_All) #36673  18
```
#assign group, normalize, design matrix
#Fisrt consider the different facilities as a batch, to check how it changes after removing batch effect

```{r}
samples<- data.frame(file=colnames(counts_All),                     facility=factor(sub("(B|D|N)(S|F|L|R)(2|3|6)","\\1",colnames(counts_All))),
                      trt=factor(sub("(B|D|N)(S|F|L|R)(2|3|6)","\\2",colnames(counts_All))),
                    
                      genotype=factor(sub("(B|D|N)(S|F|L|R)(2|3|6)","\\3",colnames(counts_All)))) 
head(samples)
save(samples,file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/DEgenes_Analysis/Ferula_RNAseq_combined_drap_oases_plant6_samples_WGCNA_kalliso.Rdata")
```

#assign group by combining all the experimental factors into one combined factor

```{r}
genotype<-levels(samples$genotype)
samples$group <- paste(samples$genotype,samples$trt,samples$facility,sep=".")
samples$genotype<-as.character(samples$genotype)
```

#edgeR

```{r}
library(edgeR)
dge <- DGEList(counts=counts_All, group=samples$group) 
length(colnames(counts_All)) # 18
dge<-calcNormFactors(dge, method = "TMM")
# look at the normalization factors
nrow(dge$samples) # 18 
hist(dge$samples[,3]) 
plot(log10(dge$sample[,"lib.size"]),dge$sample[,"norm.factors"]) 
#when we want to change the ref from first icon to others
samples$genotype <- as.factor(samples$genotype)
samples$genotype <- relevel(samples$genotype,ref="3") # plant3 reads including all three facilities
samples$trt <- as.factor(samples$trt)
samples$trt <- relevel(samples$trt,ref="R")
#design model for each factor (genotype & trt factors)
design1 <- model.matrix(~genotype+trt + facility, data=samples)
colnames(design1)
#First the overall dispersion
dge <- estimateGLMCommonDisp(dge,design1, verbose = T) #Disp =0.426 , BCV =0.653
dge <- estimateGLMTrendedDisp(dge,design1)
dge <- estimateGLMTagwiseDisp(dge,design1)
save(dge,file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/DEgenes_Analysis/Kallisto_drap_oases_plant6.dge.Rdata")
plotBCV(dge) 
mds.dge <- plotMDS(dge, method = "bcv",labels = dge$samples$group)
# see obvious batch effect, so Second, according to the help text for removeBatchEffect, you're not supposed to include the batch effect in the design matrix. The design matrix should include all the experimental variables, while the batch variable should indicate the technical batching.
```

#design model without considering facilities as a batch, just find DE genes on one facility

```{r}
#based subset just Davis libraries
#counts_Davis is after removing low counts read
counts_Davis <- counts_All[,c("DS6", "DF6","DL6","DR3", "DS3", "DF3" ,"DL3", "DR2", "DS2" ,"DF2" ,"DL2", "DR6")]
#colnames(counts_Davis) <- c("S6", "F6","L6","R3", "S3", "F3" ,"L3", "R2", "S2" ,"F2" ,"L2", "R6")
dim(counts_Davis) #36673 12
write.csv(counts_Davis, file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/DEgenes_Analysis/Ferula_RNAseq_combined_drap_oases_plant6_Davis_small_counts_kallisto.csv")
```

```{r}
samples_Davis_small <- data.frame(file=colnames(counts_Davis),                     
                      trt=factor(sub("(S|F|L|R)(2|3|6)","\\1",colnames(counts_Davis))),
                    
                      genotype=factor(sub("(S|F|L|R)(2|3|6)","\\2",colnames(counts_Davis)))) 

head(samples_Davis_small)
save(samples_Davis_small,file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/DEgenes_Analysis/Ferula_RNAse_combined_drap_oases_plant6_samples_Davis_small_kalliso.Rdata")
```

#assign group by combining all the experimental factors into one combined factor

```{r}
genotype<-levels(samples_Davis_small$genotype)
samples_Davis_small$group <- paste(samples_Davis_small$genotype,samples_Davis_small$trt,sep=".")
samples_Davis_small$genotype<-as.character(samples_Davis_small$genotype)
```

#edgeR

```{r}
library(edgeR)
dge <- DGEList(counts=counts_Davis, group=samples_Davis_small$group) 
length(colnames(counts_Davis)) # 12
dge<-calcNormFactors(dge, method = "TMM")
# look at the normalization factors
nrow(dge$samples) # 12 
hist(dge$samples[,3]) 
plot(log10(dge$sample[,"lib.size"]),dge$sample[,"norm.factors"]) 
#when we want to change the ref from first icon to others
samples_Davis_small$genotype <- as.factor(samples_Davis_small$genotype)
samples_Davis_small$genotype <- relevel(samples_Davis_small$genotype,ref="D3") #  
samples_Davis_small$trt <- as.factor(samples_Davis_small$trt)
samples_Davis_small$trt <- relevel(samples_Davis_small$trt,ref="DR")
#design model for each factor (genotype & trt factors)
design1 <- model.matrix(~genotype+trt, data=samples_Davis_small)
colnames(design1)
#First the overall dispersion
dge <- estimateGLMCommonDisp(dge,design1, verbose = T) #Disp =0.69 , BCV =0.83
dge <- estimateGLMTrendedDisp(dge,design1)
dge <- estimateGLMTagwiseDisp(dge,design1)
save(dge,file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/DEgenes_Analysis/Kallisto_drap_oases_plant6_Davis.dge.Rdata")
plotBCV(dge)
mds.dge <- plotMDS(dge, method = "bcv",labels = dge$samples$group)
```

#To find genes that are differentially expressed in gt 2 & 6 vs 3

```{r}
dge.fit <- glmFit(dge, design1)
colnames(dge.fit)
dge.lrt <- glmLRT(dge.fit,coef = c("genotypeD2","genotypeD6"))
#the top 10 most differentially expressed genes
topTags(dge.lrt)
summary(decideTestsDGE(dge.lrt,p=0.05))
#Extract genes with a FDR < 0.01 (could also use 0.05)
DEgenes1 <- topTags(dge.lrt,n = Inf)$table[topTags(dge.lrt,n = Inf)$table$FDR<0.05,]
dim(DEgenes1) #2413  6
colnames(DEgenes1)
#head(DEgenes1)
#save 
write.csv(DEgenes1,file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/DEgenes_Analysis/Ferula_RNAseq_drap_oases_plant6_Davis_DEgenes1.csv")
#To find genes that are differentially expressed in gt 2 & 6 vs 3 seperately
dge.lrt.gt2 <- glmLRT(dge.fit,coef = c("genotypeD2"))
topTags(dge.lrt.gt2)
summary(decideTestsDGE(dge.lrt.gt2, p=0.05))
dge.lrt.gt6 <- glmLRT(dge.fit,coef = c("genotypeD6"))
topTags(dge.lrt.gt6)
summary(decideTestsDGE(dge.lrt.gt6, p=0.05))
DEGs1 <- data.frame ("gt2" = as.data.frame(summary(decideTestsDGE(dge.lrt.gt2, p=0.05)))$Freq,
                                     "gt6" = as.data.frame(summary(decideTestsDGE(dge.lrt.gt6, p=0.05)))$Freq)
rownames(DEGs1) <- c("down", "no", "up")
DEGs1 <- DEGs1[c("down", "up"),]
DEGs1
save(DEGs1, file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/DEgenes_Analysis/Ferula_RNAseq_drap_oases_plant6_Davis_DEGs1.Rdata")
```

```{r}
library(reshape2)
library(ggplot2)
DEGs1.melt <- melt(DEGs1)
DEGs1.melt$DE <- rep(c("down", "up"), 2)
colnames(DEGs1.melt) <- c("genotype", "number", "DE")
DEGs1.melt
# reorder: up 1st down 2nd 
DEGs1.melt$DE <- factor(DEGs1.melt$DE, levels = c("up", "down"))
DEGs1.melt <- DEGs1.melt[order(DEGs1.melt$DE),]
DEGs1.melt
DEGs1.melt$gt <- gsub("(X)(\\.)(S|L|F)(\\.)", "\\1",DEGs1.melt$genotype)
DEGs1.melt$trt <- gsub("(X)(\\.)(S|L|F)(\\.)", "\\2",DEGs1.melt$genotype)
````

#making ggplot for DEGs

```{r}
library(ggplot2)
p.DEGs1 <- ggplot(data = DEGs1.melt)
p.DEGs1 <- p.DEGs1 + geom_bar(mapping = aes(fill=DE, x = factor(DE), y = number) , stat= "identity")
p.DEGs1 <- p.DEGs1 + facet_grid(~genotype) 
p.DEGs1 <- p.DEGs1 + labs(y = "number of differentially expressed genes", x = "")
p.DEGs1
ggsave(p.DEGs1, file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/DEgenes_Analysis/Ferula_RNAseq_drap_oases_plant6_Davis_p.DEGs1.png")
```

#To find genes that are differentially expressed in trt S, F & L vs R 

```{r}
dge.lrt.trt <- glmLRT(dge.fit,coef = c("trtDF","trtDL", "trtDS"))
#the top 10 most differentially expressed genes
topTags(dge.lrt.trt)
summary(decideTestsDGE(dge.lrt.trt,p=0.05))
#Extract genes with a FDR < 0.01 (could also use 0.05)
DEgenes2 <- topTags(dge.lrt.trt,n = Inf)$table[topTags(dge.lrt.trt,n = Inf)$table$FDR<0.05,]
write.csv(DEgenes2,file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/DEgenes_Analysis/Ferula_RNAseq_drap_oases_plant6_Davis_DEgenes2.csv")
dim(DEgenes2) #3346  7
colnames(DEgenes2)
head(DEgenes2)
write.table(DEgenes2,"/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/DEgenes_Analysis/Ferula_RNAseq_drap_oases_plant6_DEgenes2ID.txt", col.names = T, row.names = T)
```

##To find genes that are differentially expressed in trt S, F & L vs R seperately

```{r}
dge.lrt.trtF <- glmLRT(dge.fit,coef = c("trtDF"))
topTags(dge.lrt.trtF)
summary(decideTestsDGE(dge.lrt.trtF, p=0.05))
dge.lrt.trtL <- glmLRT(dge.fit,coef = c("trtDL"))
topTags(dge.lrt.trtL)
summary(decideTestsDGE(dge.lrt.trtL, p=0.05))
dge.lrt.trtS <- glmLRT(dge.fit,coef = c("trtDS"))
topTags(dge.lrt.trtS)
summary(decideTestsDGE(dge.lrt.trtS, p=0.05))
```

```{r}
#For finding the expression of each genes for each tissues
a <- decideTestsDGE(dge.lrt.trtF, p=0.05)
#a[1:3]
#a[1:3,]
#a[1:3,1]
dim(a) #36673 1
b <- rownames(dge.lrt.trtF$coefficients)
c <- data.frame(b,a)
dge.lrt.trtL <- glmLRT(dge.fit,coef = c("trtDL"))
d <- decideTestsDGE(dge.lrt.trtL, p=0.05)
e <- decideTestsDGE(dge.lrt.trtS, p=0.05)
c <- data.frame(b,a,d,e)
all_trts <- c
colnames(all_trts) <- c("Gene","Flower","Leaf","Stem")
all_trts <- all_trts[,2:4]
write.table(all_trts, "/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/all_trts_kallisto_drap_oases_plant6_Davis_DEgenes2.txt", col.names = T, row.names = T)

#For checking the genes without three zero run this command then we want to see DEGs are located in which modules.
#grep -v "0 0 0" all_trts_kallisto_drap_oases_plant6_Davis_DEgenes2.txt > all_trts_kallisto_drap_oases_plant6_Davis_DEgenes2_NoThreeZero.txt 
```

#draw figure for DE genes across different tissues

```{r}
DEGs2 <- data.frame ("trtF" = as.data.frame(summary(decideTestsDGE(dge.lrt.trtF, p=0.05)))$Freq,
                                     "trtL" = as.data.frame(summary(decideTestsDGE(dge.lrt.trtL, p=0.05)))$Freq,
                                     "trtS"= as.data.frame( summary(decideTestsDGE(dge.lrt.trtS, p=0.05)))$Freq)
rownames(DEGs2) <- c("down", "no", "up")
DEGs2 <- DEGs2[c("down", "up"),]
DEGs2
save(DEGs2, file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/Ferula_RNAseq_drap_oases_plant6_DEGs2.Rdata")
DEGs2.melt <- melt(DEGs2)
DEGs2.melt$DE <- rep(c("down", "up"), 3)
colnames(DEGs2.melt) <- c("genotype", "number", "DE")
DEGs2.melt
# reorder: up 1st down 2nd 
DEGs2.melt$DE <- factor(DEGs2.melt$DE, levels = c("up", "down"))
DEGs2.melt <- DEGs2.melt[order(DEGs2.melt$DE),]
DEGs2.melt
DEGs2.melt$gt <- gsub("(X)(\\.)(S|L|F)(\\.)", "\\1",DEGs2.melt$genotype)
DEGs2.melt$trt <- gsub("(X)(\\.)(S|L|F)(\\.)", "\\2",DEGs2.melt$genotype)
### making ggplot for DEGs
library(ggplot2)
p.DEGs2 <- ggplot(data = DEGs2.melt)
p.DEGs2 <- p.DEGs2 + geom_bar(mapping = aes(fill=DE, x = factor(DE), y = number) , stat= "identity")
p.DEGs2 <- p.DEGs2 + facet_grid(~genotype) 
p.DEGs2 <- p.DEGs2 + labs(y = "number of differentially expressed genes", x = "")
p.DEGs2
ggsave(p.DEGs2, file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/Ferula_RNAseq_drap_oases_plant6_Davis_p.DEGs2.png")
```

#edgeR for other design 

```{r}
library(edgeR)
dge <- DGEList(counts=counts_Davis, group=samples_Davis_small$group) 
length(colnames(counts_Davis)) # 12
dge<-calcNormFactors(dge, method = "TMM")
# look at the normalization factors
nrow(dge$samples) # 12 
hist(dge$samples[,3]) 
plot(log10(dge$sample[,"lib.size"]),dge$sample[,"norm.factors"]) 
#when we want to change the ref from first icon to others
samples_Davis_small$genotype <- as.factor(samples_Davis_small$genotype)
samples_Davis_small$genotype <- relevel(samples_Davis_small$genotype,ref="D3") #  
samples_Davis_small$trt <- as.factor(samples_Davis_small$trt)
samples_Davis_small$trt <- relevel(samples_Davis_small$trt,ref="DR")
#design model for each factor (trt factors)
design1 <- model.matrix(~trt, data=samples_Davis_small)
colnames(design1)
#First the overall dispersion
dge <- estimateGLMCommonDisp(dge,design1, verbose = T) #Disp =0.91 , BCV =0.95
dge <- estimateGLMTrendedDisp(dge,design1)
dge <- estimateGLMTagwiseDisp(dge,design1)
save(dge,file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/DEgenes_Analysis/Kallisto_drap_oases_plant6_Davis.dge_just_trt.Rdata")
plotBCV(dge)
mds.dge <- plotMDS(dge, method = "bcv",labels = dge$samples$group)
```

#To find genes that are differentially expressed in trt S, F & L vs R 

```{r}
dge.lrt.trt <- glmLRT(dge.fit,coef = c("trtDF","trtDL", "trtDS"))
#the top 10 most differentially expressed genes
topTags(dge.lrt.trt)
summary(decideTestsDGE(dge.lrt.trt,p=0.05))
#Extract genes with a FDR < 0.01 (could also use 0.05)
DEgenes2 <- topTags(dge.lrt.trt,n = Inf)$table[topTags(dge.lrt.trt,n = Inf)$table$FDR<0.05,]
write.csv(DEgenes2,file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/DEgenes_Analysis/Ferula_RNAseq_drap_oases_plant6_Davis_just_trt_DEgenes2.csv")
dim(DEgenes2) #3279  7
colnames(DEgenes2)
head(DEgenes2)
write.table(DEgenes2,"/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/DEgenes_Analysis/Ferula_RNAseq_drap_oases_plant6_just_trt_DEgenes2ID.txt", col.names = T, row.names = T)
```

##To find genes that are differentially expressed in trt S, F & L vs R seperately

```{r}
dge.lrt.trtF <- glmLRT(dge.fit,coef = c("trtDF"))
topTags(dge.lrt.trtF)
summary(decideTestsDGE(dge.lrt.trtF, p=0.05))
dge.lrt.trtL <- glmLRT(dge.fit,coef = c("trtDL"))
topTags(dge.lrt.trtL)
summary(decideTestsDGE(dge.lrt.trtL, p=0.05))
dge.lrt.trtS <- glmLRT(dge.fit,coef = c("trtDS"))
topTags(dge.lrt.trtS)
summary(decideTestsDGE(dge.lrt.trtS, p=0.05))
```

```{r}
#For finding the expression of each genes for each tissues
a <- decideTestsDGE(dge.lrt.trtF, p=0.05)
#a[1:3]
#a[1:3,]
#a[1:3,1]
dim(a) #36673 1
b <- rownames(dge.lrt.trtF$coefficients)
c <- data.frame(b,a)
dge.lrt.trtL <- glmLRT(dge.fit,coef = c("trtDL"))
d <- decideTestsDGE(dge.lrt.trtL, p=0.05)
e <- decideTestsDGE(dge.lrt.trtS, p=0.05)
c <- data.frame(b,a,d,e)
all_trts <- c
colnames(all_trts) <- c("Gene","Flower","Leaf","Stem")
all_trts <- all_trts[,2:4]
write.table(all_trts, "/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/all_trts_kallisto_drap_oases_plant6_Davis_just_trt_DEgenes2.txt", col.names = T, row.names = T)

#For checking the genes without three zero run this command then we want to see DEGs are located in which modules.
#grep -v "0 0 0" all_trts_kallisto_drap_oases_plant6_Davis_DEgenes2.txt > all_trts_kallisto_drap_oases_plant6_Davis_DEgenes2_NoThreeZero.txt 
```

#draw figure for DE genes across different tissues

```{r}
DEGs2 <- data.frame ("trtF" = as.data.frame(summary(decideTestsDGE(dge.lrt.trtF, p=0.05)))$Freq,
                                     "trtL" = as.data.frame(summary(decideTestsDGE(dge.lrt.trtL, p=0.05)))$Freq,
                                     "trtS"= as.data.frame( summary(decideTestsDGE(dge.lrt.trtS, p=0.05)))$Freq)
rownames(DEGs2) <- c("down", "no", "up")
DEGs2 <- DEGs2[c("down", "up"),]
DEGs2
save(DEGs2, file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/Ferula_RNAseq_drap_oases_plant6_DEGs2.Rdata")
DEGs2.melt <- melt(DEGs2)
DEGs2.melt$DE <- rep(c("down", "up"), 3)
colnames(DEGs2.melt) <- c("genotype", "number", "DE")
DEGs2.melt
# reorder: up 1st down 2nd 
DEGs2.melt$DE <- factor(DEGs2.melt$DE, levels = c("up", "down"))
DEGs2.melt <- DEGs2.melt[order(DEGs2.melt$DE),]
DEGs2.melt
DEGs2.melt$gt <- gsub("(X)(\\.)(S|L|F)(\\.)", "\\1",DEGs2.melt$genotype)
DEGs2.melt$trt <- gsub("(X)(\\.)(S|L|F)(\\.)", "\\2",DEGs2.melt$genotype)
### making ggplot for DEGs
library(ggplot2)
p.DEGs2 <- ggplot(data = DEGs2.melt)
p.DEGs2 <- p.DEGs2 + geom_bar(mapping = aes(fill=DE, x = factor(DE), y = number) , stat= "identity")
p.DEGs2 <- p.DEGs2 + facet_grid(~genotype) 
p.DEGs2 <- p.DEGs2 + labs(y = "number of differentially expressed genes", x = "")
p.DEGs2
ggsave(p.DEGs2, file="/Users/hajaramini/Documents/Ferula_RNAseq_Rbased/Ferula_RNAseq_drap_oases_plant6_Davis_p.DEGs2.png")
```
