---
title: "RNASeq_Ferula_Kallisto_DEgenes_Oases_Plant6_Davis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#This is the last version of Differential expression genes

#After mapping reads in to kallisto, we want to find DEG for Drap_Oases_Plant6.fasta, this fasta are from adding transcripts which are not available in drap assembled transcriptome of Drap result in to drap assembly  
#import data

```{r}
counts_All <- read.csv("~/Ferula_RNAseq_Analysis/Kallisto_out/combined_drap_oases_plant6_counts_kallisto.csv", header=T, row.names = "target_id")
head(counts_All) #for pooling two library of BR3 (they are not replication)
counts_All$BR3.2 <- counts_All$BR3 + counts_All$BR3.1
counts_All$BR3 <- NULL
counts_All$BR3.1 <- NULL
colnames(counts_All)[19] <- "BR3"
counts_All$BF3.2 <- counts_All$BF3 + counts_All$BF3.1
counts_All$BF3 <- NULL
counts_All$BF3.1 <- NULL
colnames(counts_All)[18] <- "BF3"
colnames(counts_All)[1] 
rownames(counts_All)
dim(counts_All) # 60134  18 ,60134  is num of transcripts in Reference
write.csv(counts_All, file="~/Ferula_RNAseq_Analysis/Kallisto_out/Ferula_RNAseq_combined_drap_oases_plant6_counts_kallisto.csv")
counts_Davis <- counts_All[,c("DS6", "DF6","DL6","DR3", "DS3", "DF3" ,"DL3", "DR2", "DS2" ,"DF2" ,"DL2", "DR6")]
colnames(counts_Davis) <- c("S6", "F6","L6","R3", "S3", "F3" ,"L3", "R2", "S2" ,"F2" ,"L2", "R6")
dim(counts_Davis) #60134 12
counts_Davis <- counts_Davis[,c("R2","R3","R6","S2","S3","S6","F2","F3","F6","L2","L3","L6")] #change the order or colnames to put same tissues near each other
write.csv(counts_Davis, file="~/Ferula_RNAseq_Analysis/Kallisto_out/Ferula_RNAseq_combined_drap_oases_plant6_Davis_counts_kallisto.csv")
```

#filter based on read count, assign group, normalize, design matrix

```{r}
colSums(counts_Davis)
hist(colSums(counts_Davis,na.rm=TRUE))

#general threshold
colSums(counts_Davis,na.rm=TRUE) > 1000000 #just false for DF3
counts_Davis_nolow <- counts_Davis[,colSums(counts_Davis,na.rm=TRUE) > 1000000]
dim(counts_Davis_nolow) #60134 11 but I prefer to work with all 18 libraries
#sample description
samples_Davis <- data.frame(file=colnames(counts_Davis),
                    trt=factor(sub("(S|F|L|R)(2|3|6)","\\1",colnames(counts_Davis))),
                    
                      genotype=factor(sub("(S|F|L|R)(2|3|6)","\\2",colnames(counts_Davis)))) 
head(samples_Davis) 

#convert NA to zero
counts_Davis[is.na(counts_Davis)]<-0
# eliminating genes with low expression levels by retaining genes with > 10 reads in > 3 samples
counts_Davis_small <-counts_Davis[rowSums(counts_Davis > 10) >= 2,] 
dim(counts_Davis_small) #27694   12
dim(counts_Davis) #60134
write.csv(counts_Davis_small, file="~/Ferula_RNAseq_Analysis/Kallisto_out/Ferula_RNAseq_combined_drap_oases_plant6_Davis_small_counts_kallisto.csv") # use this for downstream analysis

```

```{r}
samples_Davis_small <- data.frame(file=colnames(counts_Davis_small),                     
                      trt=factor(sub("(R|S|F|L)(2|3|6)","\\1",colnames(counts_Davis_small))),
                    
                      genotype=factor(sub("(R|S|F|L)(2|3|6)","\\2",colnames(counts_Davis_small)))) 
head(samples_Davis_small)
save(samples_Davis_small,file="~/Ferula_RNAseq_Analysis/Kallisto_out/Ferula_RNAse_combined_drap_oases_plant6_samples_Davis_small_kalliso.Rdata")
```

#assign group by combining all the experimental factors into one combined factor

```{r}
genotype<-levels(samples_Davis_small$genotype)
samples_Davis_small$group <- paste(samples_Davis_small$genotype,samples_Davis_small$trt,sep=".")
samples_Davis_small$genotype<-as.character(samples_Davis_small$genotype)
```

#edgeR

```{r}
library(edgeR)
dge <- DGEList(counts=counts_Davis_small, group=samples_Davis_small$group) 
length(colnames(counts_Davis_small)) # 12
dge<-calcNormFactors(dge, method = "TMM")
# look at the normalization factors
nrow(dge$samples) # 12 
hist(dge$samples[,3]) 
plot(log10(dge$sample[,"lib.size"]),dge$sample[,"norm.factors"]) 
#when we want to change the ref from first icon to others
samples_Davis_small$genotype <- as.factor(samples_Davis_small$genotype)
samples_Davis_small$genotype <- relevel(samples_Davis_small$genotype,ref="3") # plant3 reads 
samples_Davis_small$trt <- as.factor(samples_Davis_small$trt)
samples_Davis_small$trt <- relevel(samples_Davis_small$trt,ref="R")
#design model for each factor (genotype & trt factors)
design1 <- model.matrix(~genotype+trt, data=samples_Davis_small)
colnames(design1)
#First the overall dispersion
dge <- estimateGLMCommonDisp(dge,design1, verbose = T) #Disp =0.66 , BCV =0.81
dge <- estimateGLMTrendedDisp(dge,design1)
dge <- estimateGLMTagwiseDisp(dge,design1)
save(dge,file="~/Ferula_RNAseq_Analysis/Kallisto_drap_oases_plant6_Davis.dge.Rdata")
plotBCV(dge)
mds.dge <- plotMDS(dge, method = "bcv",labels = dge$samples$group)
```

#To find genes that are differentially expressed in gt 2 & 6 vs 3

```{r}
dge.fit <- glmFit(dge, design1)
colnames(dge.fit)
dge.lrt <- glmLRT(dge.fit,coef = c("genotype2","genotype6"))
#the top 10 most differentially expressed genes
topTags(dge.lrt)
summary(decideTestsDGE(dge.lrt,p=0.05))
#Extract genes with a FDR < 0.01 (could also use 0.05)
DEgenes1 <- topTags(dge.lrt,n = Inf)$table[topTags(dge.lrt,n = Inf)$table$FDR<0.05,]
dim(DEgenes1) # 2438   6
colnames(DEgenes1)
#head(DEgenes1)
#save 
write.csv(DEgenes1,file="~/Ferula_RNAseq_Analysis/Kallisto_out/Ferula_RNAseq_drap_oases_plant6_Davis_DEgenes1.csv")
#To find genes that are differentially expressed in gt 2 & 6 vs 3 seperately
dge.lrt.gt2 <- glmLRT(dge.fit,coef = c("genotype2"))
topTags(dge.lrt.gt2)
summary(decideTestsDGE(dge.lrt.gt2, p=0.05))
dge.lrt.gt6 <- glmLRT(dge.fit,coef = c("genotype6"))
topTags(dge.lrt.gt6)
summary(decideTestsDGE(dge.lrt.gt6, p=0.05))
DEGs1 <- data.frame ("gt2" = as.data.frame(summary(decideTestsDGE(dge.lrt.gt2, p=0.05)))$Freq,
                                     "gt6" = as.data.frame(summary(decideTestsDGE(dge.lrt.gt6, p=0.05)))$Freq)
rownames(DEGs1) <- c("down", "no", "up")
DEGs1 <- DEGs1[c("down", "up"),]
DEGs1
save(DEGs1, file="~/Ferula_RNAseq_Analysis/Kallisto_out/Ferula_RNAseq_drap_oases_plant6_Davis_DEGs1.Rdata")
```

```{r}
library(reshape2)
library(ggplot2)
DEGs1.melt <- melt(DEGs1)
DEGs1.melt$DE <- rep(c("down", "up"), 2)
colnames(DEGs1.melt) <- c("genotype", "number", "DE")
DEGs1.melt
# reorder: up 1st down 2nd 
DEGs1.melt$DE <- factor(DEGs1.melt$DE, levels = c("up", "down"))
DEGs1.melt <- DEGs1.melt[order(DEGs1.melt$DE),]
DEGs1.melt
DEGs1.melt$gt <- gsub("(X)(\\.)(S|L|F)(\\.)", "\\1",DEGs1.melt$genotype)

DEGs1.melt$trt <- gsub("(X)(\\.)(S|L|F)(\\.)", "\\2",DEGs1.melt$genotype)
```

#making ggplot for DEGs

```{r}
library(ggplot2)
p.DEGs1 <- ggplot(data = DEGs1.melt)
p.DEGs1 <- p.DEGs1 + geom_bar(mapping = aes(fill=DE, x = factor(DE), y = number) , stat= "identity")
p.DEGs1 <- p.DEGs1 + facet_grid(~genotype) 
p.DEGs1 <- p.DEGs1 + labs(y = "number of differentially expressed genes", x = "")
p.DEGs1
ggsave(p.DEGs1, file="~/Ferula_RNAseq_Analysis/Kallisto_out/Ferula_RNAseq_drap_oases_plant6_Davis_p.DEGs1.png")
```

#To find genes that are differentially expressed in trt S, F & L vs R 

```{r}
dge.lrt.trt <- glmLRT(dge.fit,coef = c("trtF","trtL", "trtS"))
#the top 10 most differentially expressed genes
topTags(dge.lrt.trt)
summary(decideTestsDGE(dge.lrt.trt,p=0.05))
#Extract genes with a FDR < 0.01 (could also use 0.05)
DEgenes2 <- topTags(dge.lrt.trt,n = Inf)$table[topTags(dge.lrt.trt,n = Inf)$table$FDR<0.05,]
write.csv(DEgenes2,file="~/Ferula_RNAseq_Analysis/Kallisto_out/Ferula_RNAseq_drap_oases_plant6_Davis_DEgenes2.csv")
dim(DEgenes2) #3823  7
colnames(DEgenes2)
head(DEgenes2)
write.table(DEgenes2, "~/Ferula_RNAseq_Analysis/Kallisto_out/Ferula_RNAseq_drap_oases_plant6_DEgenes2ID.txt", col.names = T, row.names = T)
```

##To find genes that are differentially expressed in trt S, F & L vs R seperately

```{r}
dge.lrt.trtF <- glmLRT(dge.fit,coef = c("trtF"))
topTags(dge.lrt.trtF)
summary(decideTestsDGE(dge.lrt.trtF, p=0.05))
dge.lrt.trtL <- glmLRT(dge.fit,coef = c("trtL"))
topTags(dge.lrt.trtL)
summary(decideTestsDGE(dge.lrt.trtL, p=0.05))
dge.lrt.trtS <- glmLRT(dge.fit,coef = c("trtS"))
topTags(dge.lrt.trtS)
summary(decideTestsDGE(dge.lrt.trtS, p=0.05))
```

```{r}
#For finding the expression of each genes for each tissues
a <- decideTestsDGE(dge.lrt.trtF, p=0.05)
#a[1:3]
#a[1:3,]
#a[1:3,1]
dim(a) #27694   1
b <- rownames(dge.lrt.trtF$coefficients)
c <- data.frame(b,a)
dge.lrt.trtL <- glmLRT(dge.fit,coef = c("trtL"))
d <- decideTestsDGE(dge.lrt.trtL, p=0.05)
e <- decideTestsDGE(dge.lrt.trtS, p=0.05)
c <- data.frame(b,a,d,e)
all_trts <- c
colnames(all_trts) <- c("Gene","Flower","Leaf","Stem")
write.table(all_trts, "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/Kallisto_out/all_trts_kallisto_drap_oases_plant6_Davis_DEgenes2.txt", col.names = T, row.names = T)

#For checking the genes without three zero run this command then we want to see DEGs are located in which modules.
#grep -v "0 0 0" all_trts_kallisto_drap_oases_plant6_Davis_DEgenes2.txt > all_trts_kallisto_drap_oases_plant6_Davis_DEgenes2_NoThreeZero.txt 
```

#draw figure for DE genes across different tissues

```{r}
DEGs2 <- data.frame ("trtF" = as.data.frame(summary(decideTestsDGE(dge.lrt.trtF, p=0.05)))$Freq,
                                     "trtL" = as.data.frame(summary(decideTestsDGE(dge.lrt.trtL, p=0.05)))$Freq,
                                     "trtS"= as.data.frame( summary(decideTestsDGE(dge.lrt.trtS, p=0.05)))$Freq)
rownames(DEGs2) <- c("down", "no", "up")
DEGs2 <- DEGs2[c("down", "up"),]
DEGs2
save(DEGs2, file="~/Ferula_RNAseq_Analysis/Kallisto_out/Ferula_RNAseq_drap_oases_plant6_DEGs2.Rdata")
DEGs2.melt <- melt(DEGs2)
DEGs2.melt$DE <- rep(c("down", "up"), 3)
colnames(DEGs2.melt) <- c("genotype", "number", "DE")
DEGs2.melt
# reorder: up 1st down 2nd 
DEGs2.melt$DE <- factor(DEGs2.melt$DE, levels = c("up", "down"))
DEGs2.melt <- DEGs2.melt[order(DEGs2.melt$DE),]
DEGs2.melt
DEGs2.melt$gt <- gsub("(X)(\\.)(S|L|F)(\\.)", "\\1",DEGs2.melt$genotype)
DEGs2.melt$trt <- gsub("(X)(\\.)(S|L|F)(\\.)", "\\2",DEGs2.melt$genotype)
### making ggplot for DEGs
library(ggplot2)
p.DEGs2 <- ggplot(data = DEGs2.melt)
p.DEGs2 <- p.DEGs2 + geom_bar(mapping = aes(fill=DE, x = factor(DE), y = number) , stat= "identity")
p.DEGs2 <- p.DEGs2 + facet_grid(~genotype) 
p.DEGs2 <- p.DEGs2 + labs(y = "number of differentially expressed genes", x = "")
p.DEGs2
ggsave(p.DEGs2, file="~/Ferula_RNAseq_Analysis/Kallisto_out/Ferula_RNAseq_drap_oases_plant6_Davis_p.DEGs2.png")
```


```{r}
source("https://bioconductor.org/biocLite.R")
biocLite("DESeq2")
library("DESeq2")
#library("lattice")
#library(tximport)
#library(readr)
#library(gplots)
#library(RColorBrewer)
#library(matrixStats)
dds.dge <- DESeqDataSetFromMatrix(countData = round(counts_Davis_small), colData = samples_Davis_small, design = ~genotype+trt)
vsd.dge <- varianceStabilizingTransformation(dds.dge)
vstMat.dge <- assay(vsd.dge)
colnames(vstMat.dge) <- colnames(counts_Davis_small)
```
# clustering to check sample seperation 

```{r}
#install.packages("ggdendro")
library(ggdendro)
#install.packages("pvclust")
library(pvclust)
hc.mds.dge <- hclust(dist(t(vstMat.dge[1:1000,])))
plot(hc.mds.dge)
rect.hclust(hc.mds.dge, k = 4, border = "red")

ggdata <- dendro_data(as.dendrogram(hc.mds.dge))
ggdata$labels$gt <- gsub("(R|S|F|L)(2|3|6)","\\2" , ggdata$labels$label) 
ggdata$labels$trt <- gsub("(R|S|F|L)(2|3|6)","\\1", ggdata$labels$label)

ggdata$labels$group <- paste(ggdata$labels$gt, ggdata$labels$trt, sep = "_")
ggdata$labels

# start ggplot here 
library(ggplot2)
p2 <- ggplot(data = segment(ggdata))
p2 <- p2 + geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) + theme_dendro()
p2 <- p2 + geom_text(data = label(ggdata), aes(x = x, y = y, label = label, hjust=0, color=group)) + coord_flip() + scale_y_reverse(expand=c(0.2, 0))
p2
ggsave("~/Ferula_RNAseq_Analysis/Kallisto_out/clustering_vst.png", width = 13, height = 25)
```
# 1) trasform your dataset from wide to long format

```{r}
library(reshape2)
vstMat.dge.melt <- melt(vstMat.dge)
head(vstMat.dge.melt)
tail(vstMat.dge.melt)
dim(vstMat.nolow.1.melt) # something wrong? 
```
# 2) get description for each factor of your VST normalized read count

```{r}
vstMat.dge.melt$genotype<-factor(gsub("(R|S|F|L)(2|3|6)","\\2",vstMat.dge.melt$Var2))
vstMat.dge.melt$trt<-factor(gsub("(R|S|F|L)(2|3|6)","\\1",vstMat.dge.melt$Var2))
```

# 3) get the mean across biological replicate

```{r}
vstMat.dge.melt.mean <- as.data.frame(tapply(vstMat.dge.melt$value,list(vstMat.dge.melt$Var1,vstMat.dge.melt$genotype,vstMat.dge.melt$trt),mean)) #this way, we get the mean expression values from 3 biological reps for each condition 
save(vstMat.dge.melt.mean, file = "~/Ferula_RNAseq_Analysis/Kallisto_out/vstMat.dge.melt.mean.Rdata")
head(vstMat.dge.melt.mean) 
```

# 4) calculate the fold change 

```{r}
vstMat.dge.log2response.FvsR<-as.data.frame(vstMat.dge.melt.mean[,grep("F",colnames(vstMat.dge.melt.mean))] - vstMat.dge.melt.mean[,grep("R",colnames(vstMat.dge.melt.mean))]) # F/R (this is log2 fold change)
vstMat.log2response.FvsR<-sapply(vstMat.dge.log2response.FvsR,as.numeric) # as numeric

vstMat.dge.log2response.SvsR<-as.data.frame(vstMat.dge.melt.mean[,grep("S",colnames(vstMat.dge.melt.mean))] - vstMat.dge.melt.mean[,grep("R",colnames(vstMat.dge.melt.mean))]) # S/R (this is log2 fold change)
vstMat.log2response.SvsR <-sapply(vstMat.dge.log2response.SvsR,as.numeric) # as numeric

vstMat.dge.log2response.LvsR<-as.data.frame(vstMat.dge.melt.mean[,grep("L",colnames(vstMat.dge.melt.mean))] - vstMat.dge.melt.mean[,grep("R",colnames(vstMat.dge.melt.mean))]) # L/R (this is log2 fold change)
vstMat.log2response.LvsR <-sapply(vstMat.dge.log2response.LvsR,as.numeric) # as numeric

rownames(vstMat.dge.log2response.FvsR) <- rownames(vstMat.dge.melt.mean)
head(vstMat.dge.log2response.FvsR)
colnames(vstMat.dge.log2response.FvsR) <- sub(".F","",colnames(vstMat.dge.log2response.FvsR)) # remove "F" from col names
save(vstMat.dge.log2response.FvsR,file="~/Ferula_RNAseq_Analysis/Kallisto_out/vstMat.dge.log2response.FvsR.Rdata")

rownames(vstMat.dge.log2response.SvsR) <- rownames(vstMat.dge.melt.mean)
head(vstMat.dge.log2response.SvsR)
colnames(vstMat.dge.log2response.SvsR) <- sub(".S","",colnames(vstMat.dge.log2response.FvsR)) # remove "S" from col names
save(vstMat.dge.log2response.SvsR,file="~/Ferula_RNAseq_Analysis/Kallisto_out/vstMat.dge.log2response.SvsR.Rdata")

rownames(vstMat.dge.log2response.LvsR) <- rownames(vstMat.dge.melt.mean)
head(vstMat.dge.log2response.LvsR)
colnames(vstMat.dge.log2response.LvsR) <- sub(".L","",colnames(vstMat.dge.log2response.FvsR)) # remove "L" from col names
save(vstMat.dge.log2response.LvsR,file="~/Ferula_RNAseq_Analysis/Kallisto_out/vstMat.dge.log2response.LvsR.Rdata")
```

```{r}
vstMat.dge.log2response.FvsR.melt <- melt(vstMat.dge.log2response.FvsR)
head(vstMat.dge.log2response.FvsR.melt)
class(vstMat.dge.log2response.FvsR.melt$Var2)
vstMat.dge.log2response.FvsR.melt
vstMat.dge.log2response.FvsR.melt$Var2 <-rownames(vstMat.dge.melt.mean)
unique(vstMat.dge.log2response.FvsR.melt$Var2)

vstMat.dge.log2response.FvsR.melt$gt <- gsub("(R|S|F|L)(2|3|6)","\\2", vstMat.dge.log2response.FvsR.melt$Var2)

vstMat.dge.log2response.FvsR.melt$trt <- gsub("(R|S|F|L)(2|3|6)","\\1", vstMat.dge.log2response.FvsR.melt$Var2)
                                           `
```
