title: "RNASeq_Ferula_Kallisto_DEG_Drap_Oases_Plant6_WGCNA_v2"
---
output: 
  html_document: 
    keep_md: yes
---

#Goal of this script is to build networks for Ferula assafoetida RNAseq data using WGCNA, will all the data together as a big net, then check the expression pattern of modules across different libraries.
#Then notice the expression is not consistent among different facilities we decide to run net just for Davis facility. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(WGCNA) 
library(tidyverse) 
library(edgeR)
library(DESeq2)
library(magrittr) 
library(gplots)
library(reshape2)
library(sva) 
options(stringsAsFactors = FALSE)
```

#run WGCNA for all genes count NOT DE genes
#import input from Kallisto_out folder

```{r}
# Filter out feature if there are less than 10 reads in more than 90% of samples (WGCNA suggestion)
#counts_Davis_WGCNA <- counts_All[rowSums(counts_All > 10) >= 3,] 
#dim(counts_Davis_WGCNA) #36673 18
#write.csv(counts_Davis_WGCNA,file="~/Ferula_RNAseq_Rbased/Kallisto_out/Ferula_RNAseq_combined_drap_oases_plant6_counts_WGCNA_kalliso.csv")
counts_WGCNA <- read.csv("~/Ferula_RNAseq_Analysis/WGCNA_out/Ferula_RNAseq_combined_drap_oases_plant6_counts_WGCNA_kalliso.csv", header = T , row.names=1)
dim(counts_WGCNA) # 36673 18
```

# Splitting the samples into different tissues types
# Normalizing function
#Transform count data to log2-counts per million (logCPM), estimate the mean-variance relationship and use this to compute appropriate observation-level weights. The data are then ready for linear modelling.

```{r}
tissuesvoom = t(voom(counts_WGCNA)$E) #all four tissues together
leafvoom = t(voom(leaf)$E)
stemvoom = t(voom(stem)$E)
rootvoom = t(voom(root)$E)
flowervoom = t(voom(flower)$E)
```

```{r}
#To check hist before and after normalization with voom function
tissuesvoom[,1] %>% hist()
counts_WGCNA$DS6 %>% hist()
counts_WGCNA$DS6[counts_All$DS6 != 0] %>% min()
```

# Checks for genes and samples with too many missing values, not have missing value
# We work with five sets, 
#Check the cluster of the samples 
#Plot to see what soft-thresholding powers would be appropriate

```{r}
# Choose a set of soft-thresholding powers
powers = c(seq(4,10,by=1), seq(12,20, by=2))

# Initialize a list to hold the results of scale-free analysis
powerTables = vector(mode = "list", length = nSet)

# Call the network topology analysis function for each set in turn
#The aim of pickSoftThreshold function is to help the user pick an appropriate soft-thresholding power for network construction
for (set in 1:nSet)
powerTables[[set]] = list(data = pickSoftThreshold(multiExpr[[set]]$data, powerVector=powers, verbose = 2)[[2]])
collectGarbage()

# Plot the results:
colors = c("yellow", "red", "blue", "green","black")

# Will plot these columns of the returned scale free analysis tables
plotCols = c(2,5,6,7)
colNames = c("Scale Free Topology Model Fit", "Mean connectivity", "Median connectivity",
"Max connectivity")

# Get the minima and maxima of the plotted points
ylim = matrix(NA, nrow = 2, ncol = 4);
for (set in 1:nSet) {
  for (col in 1:length(plotCols)) {
    ylim[1, col] = min(ylim[1, col], powerTables[[set]]$data[, plotCols[col]], na.rm = TRUE);
    ylim[2, col] = max(ylim[2, col], powerTables[[set]]$data[, plotCols[col]], na.rm = TRUE);
  }
}
```

# Plot the quantities in the chosen columns vs. the soft thresholding power

```{r}
sizeGrWindow(8, 6)
pdf(file = "~/Ferula_RNAseq_Analysis/WGCNA_out/Plot/scaleFreeAnalysis.v2.pdf", wi = 8, he = 6)
par(mfcol = c(2,2));
par(mar = c(4.2, 4.2 , 2.2, 0.5))
cex1 = 0.7;
for (col in 1:length(plotCols)) for (set in 1:nSet) {
  if (set==1) {
    plot(powerTables[[set]]$data[,1], -sign(powerTables[[set]]$data[,3])*powerTables[[set]]$data[,2],
        xlab="Soft Threshold (power)",ylab=colNames[col],type="n", ylim = ylim[, col],
        main = colNames[col]);
    addGrid();
  }
  if (col==1) {
    text(powerTables[[set]]$data[,1], -sign(powerTables[[set]]$data[,3])*powerTables[[set]]$data[,2],
        labels=powers,cex=cex1,col=colors[set]);
  } else
  text(powerTables[[set]]$data[,1], powerTables[[set]]$data[,plotCols[col]],
      labels=powers,cex=cex1,col=colors[set]);

  if (col==1) {
    legend("bottomright", legend = setLabels, col = colors, pch = 20) ;
  } else
    legend("topright", legend = setLabels, col = colors, pch = 20) ;
}
dev.off();

```

#Build the modules for all four tissues together
#Build module for root and flower, separately.
#Leaf and stem do not have enough samples. Try just with flower and root

```{r}
net_tissues =  blockwiseModules(
datExpr0tissues, power = 10, minModuleSize = 30, deepSplit = 2,
pamRespectsDendro = FALSE,
mergeCutHeight = 0.25, numericLabels = TRUE,
minKMEtoStay = 0,
saveTOMs = TRUE, verbose = 3)
table(net_tissues$colors) 
#0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16   17   18   19 
#80 9155 4014 3885 3414 1813 1789 1633 1321 1240 1222  972  723  718  648  604  516  468  413  242 
#20   21   22   23   24   25   26   27   28   29   30   31   32   33   34   35   36   37   38   39 
#209  178  149  143  119  106  101   97   96   91   77   71   67   59   45   43   41   39   38   34 
t <- net_tissues$colors
write.table(t, "~/Ferula_RNAseq_Analysis/WGCNA_out/tissues_modules.txt")
```

```{r}
# Convert the numeric labels to color labels
#by parsing through blockgenes, notice we have 8 block including around 4500 genes for speeding up.by mergedColors[net_tissues$blockGenes[[1]]] got just module lable for blockgenes.

mergedColors = labels2colors(net_tissues$colors)
mergedColors = labels2colors(net_root$colors)
# Plot the dendrogram and the module colors underneath
sizeGrWindow(8,6);
pdf(file = "../WGCNA_out/Plot/networkConstructionTissues.pdf", wi = 8, he = 6)
plotDendroAndColors(net_tissues$dendrograms[[1]], mergedColors[net_tissues$blockGenes[[1]]],
"Module colors",
dendroLabels = FALSE, hang = 0.03,
addGuide = TRUE, guideHang = 0.05,
main ="Net_Tissues")
dev.off()
```


```{r}
moduleLabels = net_tissues$colors
moduleColors = labels2colors(net_tissues$colors)
MEs = net_tissues$MEs
geneTree = net_tissues$dendrograms[[1]]
save(MEs, moduleLabels, moduleColors, geneTree,
file = "~/Ferula_RNAseq_Analysis/WGCNA_out/networkConstruction_tissues.RData")
```

#Build module for root and flower, separately.
#Root

```{r}
net_root =  blockwiseModules(
datExpr0root, power = 10, minModuleSize = 30, deepSplit = 2,
pamRespectsDendro = FALSE,
mergeCutHeight = 0.25, numericLabels = TRUE,
minKMEtoStay = 0,
saveTOMs = TRUE, verbose = 3)
table(net_root$colors)
#1     2     3     4     5     6     7     8     9    10    11    12    13    14    15    16    17 
#13161  6184  4944  1889  1612  1519  1425  1220  1068   898   757   541   274   272   247   213   207 
#18    19 
#164    78
r <- net_root$colors #for finding co expressed genes in each module
write.table(r, "~/Ferula_RNAseq_Analysis/WGCNA_out/root_modules.txt")
```

```{r}
# Convert the numeric labels to color labels
mergedColors = labels2colors(net_root$colors)
# Plot the dendrogram and the module colors underneath
sizeGrWindow(8,6);
pdf(file = "~/Ferula_RNAseq_Analysis/WGCNA_out/Plot/networkConstructionRoot.pdf", wi = 8, he = 6)
plotDendroAndColors(net_root$dendrograms[[1]], mergedColors[net_root$blockGenes[[1]]],
"Module colors",
dendroLabels = FALSE, hang = 0.03,
addGuide = TRUE, guideHang = 0.05,
main ="Net_Root")
dev.off()
```

```{r}
moduleLabels_root = net_root$colors
moduleColors_root = labels2colors(net_root$colors)
MEs_root = net_root$MEs
geneTree_root = net_root$dendrograms[[1]]
save(MEs_root, moduleLabels_root, moduleColors_root, geneTree_root,
file = "~/Ferula_RNAseq_Analysis/WGCNA_out/networkConstruction_root.RData")
```

#Flower

```{r}
net_flower =  blockwiseModules(
datExpr0flower, power = 10, minModuleSize = 30, deepSplit = 2,
pamRespectsDendro = FALSE,
mergeCutHeight = 0.25, numericLabels = TRUE,
minKMEtoStay = 0,
saveTOMs = TRUE, verbose = 3)
table(net_flower$colors)
#1     2     3     4     5     6     7     8     9    10    11    12    13    14    15    16    17 
#18786  5623  3177  1693  1147  1082   922   859   670   617   526   466   393   287   263    93    36 
#18 
#33
f <- net_flower$colors #for finding co expressed genes in each module
write.table(f, "~/Ferula_RNAseq_Analysis/WGCNA_out/flower_modules.txt")
```

```{r}
#use blockGenes[[2]] instead of [[1]] becuase blockGenes[[1]] got two types colours :brown turquoise 
# Convert the numeric labels to color labels
mergedColors = labels2colors(net_flower$colors)
# Plot the dendrogram and the module colors underneath
sizeGrWindow(8,6);
pdf(file = "~/Ferula_RNAseq_Analysis/WGCNA_out/Plot/networkConstructionFlower.pdf", wi = 8, he = 6)
plotDendroAndColors(net_flower$dendrograms[[2]], mergedColors[net_flower$blockGenes[[2]]],
"Module colors",
dendroLabels = FALSE, hang = 0.03,
addGuide = TRUE, guideHang = 0.05,
main ="Net_Flower")
dev.off()
```

```{r}
moduleLabels_flower = net_flower$colors
moduleColors_flower = labels2colors(net_flower$colors)
MEs_flower= net_flower$MEs
geneTree_flower = net_flower$dendrograms[[2]]
save(MEs_flower, moduleLabels_flower, moduleColors_flower, geneTree_flower,
file = "~/Ferula_RNAseq_Analysis/WGCNA_out/networkConstruction_flower.RData")
```

#then we can see the num intersection for each modules for flower and tissues as well as root and tissues


```{r}

#continue based https://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/Consensus-RelateToFemMods.pdf
# Load the data saved in the first part
lnames = load(file = "networkConstruction_flower.RData")
lnames
#[1] "MEs_flower"  "moduleLabels_flower" "moduleColors_flower" "geneTree_flower" 
lnames <- load(file = "networkConstruction_root.RData")
lnames
#[1] "MEs_root"   "moduleLabels_root" "moduleColors_root" "geneTree_root"
lnames <- load(file = "networkConstruction_tissues.RData")
lnames
#[1] "MEs" "moduleLabels" "moduleColors" "geneTree" 

#The network analysis results are represented by the variables consMEs, moduleLabels, moduleColors, and consTree. We are now ready to relate the root, flower modules to the tissues modules (big net). We calculate the overlaps of each pair of root modules, and use the Fisher’s exact test (also known as hypergeometric test) to assign a p-value to each of the pairwise overlaps.

# Isolate the module labels in the order they appear in ordered module eigengenes

rootModuleLabels = substring(names(MEs_root), 3) #19
flowerModuleLabels = substring(names(MEs_flower), 3) #18
tissueModuleLabels = substring(names(MEs), 3) 
# Convert the numeric module labels to color labels
rootModules = labels2colors(as.numeric(rootModuleLabels)) #19
flowerModules = labels2colors(as.numeric(flowerModuleLabels)) #18
tissueModules = labels2colors(as.numeric(tissueModuleLabels)) #39
# Numbers of root and consensus modules
nRootMods = length(rootModules) #19
nFlowerMods = length(flowerModules) #18
nTissueMods = length(tissueModules) #40

# Initialize tables of p-values and of the corresponding counts.
pTableRT = matrix(0, nrow = nRootMods, ncol = nTissueMods);
CountTblRT = matrix(0, nrow = nRootMods, ncol = nTissueMods);
pTableFT = matrix(0, nrow = nFlowerMods, ncol = nTissueMods);
CountTblFT = matrix(0, nrow = nFlowerMods, ncol = nTissueMods);
# Execute all pairwaise comparisons
#First for root and tissue 
for (rmod in 1:nRootMods)
for (cmod in 1:nTissueMods)
{
rootMembers = (moduleColors_root == rootModules[rmod]);
tissueMembers = (moduleColors == tissueModules[cmod]);
pTableRT[rmod, cmod] = -log10(fisher.test(rootMembers, tissueMembers, alternative = "greater")$p.value);
CountTblRT[rmod, cmod] = sum(moduleColors_root == rootModules[rmod] & moduleColors ==
tissueModules[cmod])
}

#To display the p-value and count tables in an informative way, we create a color-coded table of the intersection counts. The colors will indicate the p-value significance:
  
# Truncate p values smaller than 10^{-50} to 10^{-50}
pTableRT[is.infinite(pTableRT)] = 1.3*max(pTableRT[is.finite(pTableRT)]);
View(pTableRT) #row 19 col 40
pTableRT[pTableRT>50 ] = 50 ;
# Marginal counts (really module sizes)
rootModTotals = apply(CountTblRT, 1, sum)
tissueModTotals = apply(CountTblRT, 2, sum)
# Actual plotting
sizeGrWindow(19,10 );
pdf(file = "~/Ferula_RNAseq_Analysis/WGCNA_out/Plot/TissuesVsRootModuless.pdf", wi = 19, he = 10);
par(mfrow=c(1,1));
par(cex = 1.0);
par(mar=c(8, 10.4, 2.7, 1)+0.3);
# Use function labeledHeatmap to produce the color-coded table with all the trimmings
labeledHeatmap(Matrix = pTableRT,
xLabels = paste(" ", tissueModules),
yLabels = paste(" ", rootModules),
colorLabels = TRUE,
xSymbols = paste("Tissue ", tissueModules, ": ", tissueModTotals, sep=""),
ySymbols = paste("Root ", rootModules, ": ", rootModTotals, sep=""),
textMatrix = CountTblRT,
colors = greenWhiteRed(100)[50:100],
main = "Correspondence of Root set-specific and Tissues modules",
cex.text = 1.0, cex.lab = 1.0, setStdMargins = FALSE);
dev.off()
#Figure: Correspondence of root set-specific modules and the tissue modules. Each row of the table corresponds to one root set-specific module (labeled by color as well as text), and each column corresponds to one tissue module. Numbers in the table indicate gene counts in the intersection of the corresponding modules. Coloring of the table encodes − log(p), with p being the Fisher’s exact test p-value for the overlap of the two modules. The stronger the red color, the more significant the overlap is.


#Second for flower and tissue 
for (rmod in 1:nFlowerMods)
for (cmod in 1:nTissueMods)
{
flowerMembers = (moduleColors_flower == flowerModules[rmod]);
tissueMembers = (moduleColors == tissueModules[cmod]);
pTableFT[rmod, cmod] = -log10(fisher.test(flowerMembers, tissueMembers, alternative = "greater")$p.value);
CountTblFT[rmod, cmod] = sum(moduleColors_flower == flowerModules[rmod] & moduleColors ==
tissueModules[cmod])
}

#To display the p-value and count tables in an informative way, we create a color-coded table of the intersection counts. The colors will indicate the p-value significance:
  
# Truncate p values smaller than 10^{-50} to 10^{-50}
pTableFT[is.infinite(pTableFT)] = 1.3*max(pTableFT[is.finite(pTableFT)]);
View(pTableFT) #row 18 col 40
pTableFT[pTableFT>50 ] = 50 ;
# Marginal counts (really module sizes)
flowerModTotals = apply(CountTblFT, 1, sum)
tissueModTotals = apply(CountTblFT, 2, sum)
# Actual plotting
sizeGrWindow(19,10 );
pdf(file = "Plot/TissuesVsFlowerModuless.pdf", wi = 19, he = 10);
par(mfrow=c(1,1));
par(cex = 1.0);
par(mar=c(8, 10.4, 2.7, 1)+0.3);
# Use function labeledHeatmap to produce the color-coded table with all the trimmings
labeledHeatmap(Matrix = pTableFT,
xLabels = paste(" ", tissueModules),
yLabels = paste(" ", flowerModules),
colorLabels = TRUE,
xSymbols = paste("Tissue ", tissueModules, ": ", tissueModTotals, sep=""),
ySymbols = paste("Flower ", flowerModules, ": ", flowerModTotals, sep=""),
textMatrix = CountTblFT,
colors = greenWhiteRed(100)[50:100],
main = "Correspondence of Flower set-specific and Tissues modules",
cex.text = 1.0, cex.lab = 1.0, setStdMargins = FALSE);
dev.off()

```

#Next we calculate the principle components for visualizations(Julin's suggestion), see the mean of expression of flower, stem, root and leaf in tissue modules 

```{r}
#To see the expression pattern of all genes in all modules of tissue network (big net) in each root samples (#40 modules with 6 root samples)
PCs_Root  = moduleEigengenes((datExpr0root), colors= moduleColors) 
ME_Root   = PCs_Root$eigengenes 
dim(ME_Root) #6 40
colors = names(table(moduleColors))
#[1] "black"          "blue"           "brown"          "cyan"           "darkgreen"      "darkgrey"      
#[7] "darkmagenta"    "darkolivegreen" "darkorange"     "darkred"        "darkturquoise"  "green"         #[13] "greenyellow"    "grey"           "grey60"         "lightcyan"      "lightgreen"     "lightyellow"   
#[19] "magenta"        "midnightblue"   "orange"         "orangered4"     "paleturquoise"  "pink"          
#[25] "plum1"          "purple"         "red"            "royalblue"      "saddlebrown"    "salmon"        
#[31] "sienna3"        "skyblue"        "skyblue3"       "steelblue"      "tan"            "turquoise"     
#[37] "violet"         "white"          "yellow"         "yellowgreen" 

#save.image("tutorial.RData")

pdf("Plot/ModuleEigengeneVisualization_Root.pdf",height=6,width=6)

par(mfrow=c(1,1), mar=c(6, 5, 2, 2) + 0.1, cex=1)

ordergenes = geneTree_root$order

for (which.module in names(table(moduleColors))){
  ME = ME_Root[, paste("ME",which.module, sep="")] 
  barplot(ME, col=which.module, main="", cex.main=2, 
      ylab="eigengene expression",xlab="array sample") 
}; 

dev.off();
```


```{r}
#To see the expression pattern of all genes in all modules of tissue network (big net) in each flower samples (#19 modules with #6 flower samples)
PCs_Flower  = moduleEigengenes((datExpr0flower), colors= moduleColors) 
ME_Flower   = PCs_Flower$eigengenes 
dim(ME_Flower) #6 40
colors = names(table(moduleColors))
#[1] "black"          "blue"           "brown"          "cyan"           "darkgreen"      "darkgrey"      
#[7] "darkmagenta"    "darkolivegreen" "darkorange"     "darkred"        "darkturquoise"  "green"         #[13] "greenyellow"    "grey"           "grey60"         "lightcyan"      "lightgreen"     "lightyellow"   
#[19] "magenta"        "midnightblue"   "orange"         "orangered4"     "paleturquoise"  "pink"          
#[25] "plum1"          "purple"         "red"            "royalblue"      "saddlebrown"    "salmon"        
#[31] "sienna3"        "skyblue"        "skyblue3"       "steelblue"      "tan"            "turquoise"     
#[37] "violet"         "white"          "yellow"         "yellowgreen" 

#save.image("tutorial.RData")

pdf("Plot/ModuleEigengeneVisualization_Flower.pdf",height=6,width=6)

par(mfrow=c(1,1), mar=c(6, 5, 2, 2) + 0.1, cex=1)

ordergenes = geneTree_flower$order

for (which.module in names(table(moduleColors))){
  ME = ME_Flower[, paste("ME",which.module, sep="")] 
  barplot(ME, col=which.module, main="", cex.main=2, 
      ylab="eigengene expression",xlab="array sample") 
}; 

dev.off();
```


```{r}
#To see the expression pattern of all genes in all modules of tissue network (big net) in each stem samples (#19 modules with #3 stem samples)
PCs_Stem = moduleEigengenes((datExpr0stem), colors= moduleColors) 
ME_Stem   = PCs_Stem$eigengenes 
dim(ME_Stem) #3 40
colors = names(table(moduleColors))
#[1] "black"          "blue"           "brown"          "cyan"           "darkgreen"      "darkgrey"      
#[7] "darkmagenta"    "darkolivegreen" "darkorange"     "darkred"        "darkturquoise"  "green"         #[13] "greenyellow"    "grey"           "grey60"         "lightcyan"      "lightgreen"     "lightyellow"   
#[19] "magenta"        "midnightblue"   "orange"         "orangered4"     "paleturquoise"  "pink"          
#[25] "plum1"          "purple"         "red"            "royalblue"      "saddlebrown"    "salmon"        
#[31] "sienna3"        "skyblue"        "skyblue3"       "steelblue"      "tan"            "turquoise"     
#[37] "violet"         "white"          "yellow"         "yellowgreen" 

#save.image("tutorial.RData")

pdf("~/Ferula_RNAseq_Analysis/WGCNA_out/Plot/ModuleEigengeneVisualization_Stem.pdf",height=6,width=6)

par(mfrow=c(1,1), mar=c(6, 5, 2, 2) + 0.1, cex=1)

for (which.module in names(table(moduleColors))){
  ME = ME_Stem[, paste("ME",which.module, sep="")] 
  barplot(ME, col=which.module, main="", cex.main=2, 
      ylab="eigengene expression",xlab="array sample") 
}; 

dev.off();
```


```{r}
#To see the expression pattern of all genes in all modules of tissue network (big net) in each leaf samples (#19 modules with #3 leaf samples)
PCs_Leaf = moduleEigengenes((datExpr0leaf), colors= moduleColors) 
ME_Leaf   = PCs_Leaf$eigengenes 
dim(ME_Leaf) #3 40
colors = names(table(moduleColors))
#[1] "black"          "blue"           "brown"          "cyan"           "darkgreen"      "darkgrey"      
#[7] "darkmagenta"    "darkolivegreen" "darkorange"     "darkred"        "darkturquoise"  "green"         #[13] "greenyellow"    "grey"           "grey60"         "lightcyan"      "lightgreen"     "lightyellow"   
#[19] "magenta"        "midnightblue"   "orange"         "orangered4"     "paleturquoise"  "pink"          
#[25] "plum1"          "purple"         "red"            "royalblue"      "saddlebrown"    "salmon"        
#[31] "sienna3"        "skyblue"        "skyblue3"       "steelblue"      "tan"            "turquoise"     
#[37] "violet"         "white"          "yellow"         "yellowgreen" 

#save.image("tutorial.RData")

pdf("~/Ferula_RNAseq_Analysis/WGCNA_out/Plot/ModuleEigengeneVisualization_Leaf.pdf",height=6,width=6)

par(mfrow=c(1,1), mar=c(6, 5, 2, 2) + 0.1, cex=1)

for (which.module in names(table(moduleColors))){
  ME = ME_Leaf[, paste("ME",which.module, sep="")] 
  barplot(ME, col=which.module, main="", cex.main=2, 
      ylab="eigengene expression",xlab="array sample") 
}; 

dev.off();
```

#I want to compare the clustering the big net modules in terms of each tissue expression
#clustering big Net in terms of root's expression

```{r}
#for root
PCs1_Root    = moduleEigengenes((datExpr0root), colors= moduleColors) 
ME1_Root    = PCs1_Root$eigengenes
dim(ME1_Root) #6 40
distPC1_Root = 1-abs(cor(ME1_Root,use="p"))
distPC1_Root = ifelse(is.na(distPC1_Root), 0, distPC1_Root)

pcTree1_Root = hclust(as.dist(distPC1_Root),method="a") 
MDS1_Root  = cmdscale(as.dist(distPC1_Root),2)
colors = names(table(moduleColors))

save.image("tutorial.RData")

####
pdf("~/Ferula_RNAseq_Analysis/WGCNA_out/Plot/pcTree1_Root.pdf",height=6,width=10)
par(mfrow=c(1,1), mar=c(6, 5, 2, 2) + 0.1, cex=1)
plot(pcTree1_Root, xlab="",ylab="",main="",sub="")
dev.off()

```

#clustering big Net in terms of flower's expression

```{r}
#for flower
PCs1_Flower    = moduleEigengenes((datExpr0flower), colors= moduleColors) 
ME1_Flower    = PCs1_Flower$eigengenes
dim(ME1_Flower) #6 40
distPC1_Flower = 1-abs(cor(ME1_Flower,use="p"))
distPC1_Flower = ifelse(is.na(distPC1_Flower), 0, distPC1_Flower)

pcTree1_Flower = hclust(as.dist(distPC1_Flower),method="a") 
MDS1_Flower  = cmdscale(as.dist(distPC1_Flower),2)
colors = names(table(moduleColors))

pdf("Plot/pcTree1_Flower.pdf",height=6,width=10)
par(mfrow=c(1,1), mar=c(6, 5, 2, 2) + 0.1, cex=1)
plot(pcTree1_Flower, xlab="",ylab="",main="",sub="")
dev.off()
```

#clustering big Net in terms of leaf's expression

```{r}
#for leaf
PCs1_Leaf    = moduleEigengenes((datExpr0leaf), colors= moduleColors) 
ME1_Leaf    = PCs1_Leaf$eigengenes
dim(ME1_Leaf) #3 40
distPC1_Leaf = 1-abs(cor(ME1_Leaf,use="p"))
distPC1_Leaf = ifelse(is.na(distPC1_Leaf), 0, distPC1_Leaf)

pcTree1_Leaf = hclust(as.dist(distPC1_Leaf),method="a") 
MDS1_Leaf  = cmdscale(as.dist(distPC1_Leaf),2)
colors = names(table(moduleColors))

pdf("~/Ferula_RNAseq_Analysis/WGCNA_out/Plot/pcTree1_Leaf.pdf",height=6,width=10)
par(mfrow=c(1,1), mar=c(6, 5, 2, 2) + 0.1, cex=1)
plot(pcTree1_Leaf, xlab="",ylab="",main="",sub="")
dev.off()
```

#clustering big Net in terms of stem's expression

```{r}
#for stem
PCs1_Stem    = moduleEigengenes((datExpr0stem), colors= moduleColors) 
ME1_Stem    = PCs1_Stem$eigengenes
dim(ME1_Stem) #3 40
distPC1_Stem = 1-abs(cor(ME1_Stem,use="p"))
distPC1_Stem = ifelse(is.na(distPC1_Stem), 0, distPC1_Stem)

pcTree1_Stem = hclust(as.dist(distPC1_Stem),method="a") 
MDS1_Stem  = cmdscale(as.dist(distPC1_Stem),2)
colors = names(table(moduleColors))

pdf("~/Ferula_RNAseq_Analysis/WGCNA_out/Plot/pcTree1_Stem.pdf",height=6,width=10)
par(mfrow=c(1,1), mar=c(6, 5, 2, 2) + 0.1, cex=1)
plot(pcTree1_Stem, xlab="",ylab="",main="",sub="")
dev.off()

```

```{r}
source("https://bioconductor.org/biocLite.R")
biocLite("sva")
library(sva)
source("http://www.bioconductor.org/biocLite.R")
biocLite(c("Biobase","sva","bladderbatch","snpStats"))
library(bladderbatch)
biocLite("pamr")
library(pamr)
library(limma)
```

```{r}
edata <- exprs(object = counts_WGCNA)
edata <- exprs(object)
mod = model.matrix(~as.factor(tissue), data=edata)
```

