---
title: "RNASeq_Ferula_Kallisto_DEG_Drap_Oases_Plant6_WGCNA_v3"
output: html_document
---

#Goal of this script is to build networks for Ferula assafoetida RNAseq data using WGCNA. Difference between this version and v2 is that in this script first we remove batch effect then construct the gene network.
#for removing batch effect using Ruijuan scripts
#https://github.com/leejimmy93/KIAT/blob/master/F2/F2_network.Rmd
#considering the expression pattern across the samples, seems batch effect is not removed in correct way

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#load libraries

```{r}
library(WGCNA) 
library(tidyverse) 
library(edgeR)
library(DESeq2)
library(magrittr) 
library(gplots)
library(reshape2)
library(sva) 
options(stringsAsFactors = FALSE)
```

#run WGCNA for all genes count NOT DE genes
#import input from Kallisto_out folder

```{r}
# Filter out feature if there are less than 10 reads in more than 90% of samples (WGCNA suggestion)
#counts_Davis_WGCNA <- counts_All[rowSums(counts_All > 10) >= 3,] 
#dim(counts_Davis_WGCNA) #36673 18
#write.csv(counts_Davis_WGCNA,file="~/Ferula_RNAseq_Analysis/WGCNA_out/Ferula_RNAseq_combined_drap_oases_plant6_counts_WGCNA_kalliso.csv")
counts_WGCNA <- read.csv("~/Ferula_RNAseq_Analysis/WGCNA_out/Ferula_RNAseq_combined_drap_oases_plant6_counts_WGCNA_kalliso.csv", header = T , row.names=1)
dim(counts_WGCNA) # 36673 18
```

#To check hist before and after normalization with voom function

```{r}
tissuesvoom = t(voom(counts_WGCNA)$E) #all four tissues together
tissuesvoom[,1] %>% hist()
counts_WGCNA$DS6 %>% hist()
counts_WGCNA$DS6[counts_All$DS6 != 0] %>% min()
```

#correct batch effect

```{r}
# check vst transformed data to see is there any batch effect. 
tissuesvoom %>% dim() # 18 36673
rownames(tissuesvoom)
#[1] "DS6" "DF6" "DL6" "DR3" "DS3" "DF3" "DL3" "DR2" "DS2" "DF2" "DL2" "DR6" "NF3" "NF6" "NR3" "NR6" "BR3" "BF3"
test1 <- as.matrix(dist(tissuesvoom))
dim(test1)
#18 18
#perform the multi-dimensional scaling
test2 <- as.data.frame(cmdscale(test1))
head(test2) #now we have 2 dimensions 
dim(test2) #18  2
plot(test2)
batch_info <- read.csv("~/Ferula_RNAseq_Rbased/WGCNA_out/batch.csv") # create file for batch information
#dim 18  2
test2$sample_ID <- rownames(test2)
test2 %>% 
  left_join(batch_info, by = "sample_ID") %>% 
  ggplot() +
  geom_point(aes(x= V1 ,y= V2, color=as.factor(batch)))# bad, clear batch effect   
ggsave("MDS_All_Libraries_Batch_Effect.png",width=11,height=8,path="/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/WGCNA_out/Plot_v3")
```

# correct batch effect 

```{r}
tissuesvoom.t <- t(tissuesvoom)
tissuesvoom.t %>% dim() #36673  18
tissuesvoom.t.batch.corrected <- removeBatchEffect(tissuesvoom.t, batch = batch_info$batch)
dim(tissuesvoom.t.batch.corrected)  #36673  18
test1 <- as.matrix(dist(t(tissuesvoom.t.batch.corrected)))  
test2 <- as.data.frame(cmdscale(test1))
test2$sample_ID <- rownames(test2)
test2 %>% 
  left_join(batch_info, by = "sample_ID") %>% 
  ggplot() + 
  geom_point(aes(x = V1, y= V2, color = as.factor(batch))) # great, batch effect removed 

ggsave("MDS_All_Libraries_No_Batch_Effect.png",width=11,height=8,path="/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/WGCNA_out/Plot_v3")
save(tissuesvoom.t.batch.corrected, file = "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/WGCNA_out/tissuesvoom.t.batch.corrected.Rdata") # use this dataset for downstream analysis
```

# get genes with large stdv across 18 libraries, the reason I use this because later on co-expression file were obtained using vst transfomred data directly. 

```{r}
load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Rbased/WGCNA_out/tissuesvoom.t.batch.corrected.Rdata")
tissuesvoom.t.batch.corrected <- as.data.frame(tissuesvoom.t.batch.corrected)
tissuesvoom.t.batch.corrected$sd <- as.numeric(apply(tissuesvoom.t.batch.corrected,1,sd))  
tissuesvoom.t.batch.corrected.largeSD<-tissuesvoom.t.batch.corrected[tissuesvoom.t.batch.corrected$sd>0.5,] 
dim(tissuesvoom.t.batch.corrected.largeSD) # 36322  19  it seems the number of the genes with large stdv is not lower than total genes just 300 lower. so I used total genes for networking 
tissuesvoom.t.batch.corrected.largeSD <- tissuesvoom.t.batch.corrected.largeSD[,1:18]
save(tissuesvoom.t.batch.corrected.largeSD, file = "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/WGCNA_out/tissuesvoom.t.batch.corrected.largeSD.Rdata")
```

# Splitting the samples into different tissues types after removing batch effect
# Normalizing function
#Transform count data to log2-counts per million (logCPM), estimate the mean-variance relationship and use this to compute appropriate observation-level weights. The data are then ready for linear modelling.

```{r}
tissuesvoom.t.batch.corrected <- tissuesvoom.t.batch.corrected[,1:18]
dim(tissuesvoom.t.batch.corrected) # 36673 18
tissuesvoom.batch.corrected <- t(tissuesvoom.t.batch.corrected)
dim(tissuesvoom.batch.corrected) #18 36673
leafvoom <- t(tissuesvoom.t.batch.corrected[,c(3, 7, 11)])
dim(leafvoom) #3 36673
stemvoom <- t(tissuesvoom.t.batch.corrected[,c(1, 5, 9)])
dim(stemvoom) #3 36673
rootvoom <- t(tissuesvoom.t.batch.corrected[,c(4, 8, 12, 15, 16, 17)])
dim(rootvoom) #6 36673
flowervoom <- t(tissuesvoom.t.batch.corrected[,c(2, 6, 10, 13, 14, 18)])
dim(flowervoom) #6 36673
```

# Checks for genes and samples with too many missing values, not have missing value

```{r}
#since samples with not missing values so leafvoom, stemvoom, rootvoom and flowervoom equal to datExpr0
datExpr0tissues = as.data.frame((tissuesvoom.batch.corrected))
rownames(datExpr0tissues) #18
nGenes <- ncol(datExpr0tissues)

datExpr0root = as.data.frame((rootvoom));
rownames(datExpr0root) #number of the samples
nSamples <- nrow(datExpr0root) #6
nGenes <- ncol(datExpr0root) #36673

gsg = goodSamplesGenes(datExpr0root) 
gsg$allOK #TRUE If the last statement returns TRUE, all genes have passed the cuts

datExpr0leaf = as.data.frame((leafvoom))
nSamples <- nrow(datExpr0leaf) #3
datExpr0stem = as.data.frame((stemvoom))
nSamples <- nrow(datExpr0stem) #3
datExpr0flower = as.data.frame((flowervoom))
nSamples <- nrow(datExpr0flower) #6
#The variables datExpr0flower, datExpr0leaf, datExpr0root, datExpr0stem,  now contains the expression data ready for network analysis
save(datExpr0tissues, datExpr0flower, datExpr0leaf, datExpr0root, datExpr0stem,leafvoom, stemvoom, rootvoom, flowervoom, file = "dataInput_batch_corrected.RData")
```

# We work with five sets, 

```{r}
#since samples with not missing values so leafvoom, stemvoom, rootvoom and flowervoom equal to datExpr0, we can use multiExpr instead of datExpr0
nSets <- 5
# For easier labeling of plots, create a vector holding descriptive names of the five sets.
setLabels = c("Leaf", "Stem", "Root", "Flower", "tissues")
shortLabels = setLabels

# Form multi-set expression data
multiExpr = vector(mode = "list", length = nSets)

multiExpr[[1]] = list(data = as.data.frame(leafvoom))
names(multiExpr[[1]]$data) = colnames(leafvoom)
rownames(multiExpr[[1]]$data) = rownames(leafvoom)
nSamples <- nrow(multiExpr[[1]]$data) #3
nGenes <- ncol(multiExpr[[1]]$data) #36673

multiExpr[[2]] = list(data = as.data.frame(stemvoom))
names(multiExpr[[2]]$data) = colnames(stemvoom)
rownames(multiExpr[[2]]$data) = rownames(stemvoom)
nSamples <- nrow(multiExpr[[2]]$data) #3
nGenes <- ncol(multiExpr[[2]]$data) #36673

multiExpr[[3]] = list(data = as.data.frame(rootvoom))
names(multiExpr[[3]]$data) = colnames(rootvoom)
rownames(multiExpr[[3]]$data) = rownames(rootvoom)
nSamples <- nrow(multiExpr[[3]]$data) #6
nGenes <- ncol(multiExpr[[3]]$data) #36673

multiExpr[[4]] = list(data = as.data.frame(flowervoom))
names(multiExpr[[4]]$data) = colnames(flowervoom)
rownames(multiExpr[[4]]$data) = rownames(flowervoom)
nSamples <- nrow(multiExpr[[4]]$data) #18
nGenes <- ncol(multiExpr[[4]]$data) #36673

multiExpr[[5]] = list(data = as.data.frame(tissuesvoom.batch.corrected))
names(multiExpr[[5]]$data) = colnames(tissuesvoom.batch.corrected)
rownames(multiExpr[[5]]$data) = rownames(tissuesvoom.batch.corrected)
nSamples <- nrow(multiExpr[[5]]$data) #6
nGenes <- ncol(multiExpr[[5]]$data) #36673

# Check that the data has the correct format for many functions operating on multiple sets
exprSize = checkSets(multiExpr)
#$nSets
#[1] 5
#$nGenes
#[1] 36673
#$nSamples
#[1]  3  3  6  6 18
#$structureOK
#[1] TRUE
```

#Check the cluster of the samples

```{r}
sampleTrees = list()
for (set in 1:nSets) {
  sampleTrees[[set]] = hclust(dist(multiExpr[[set]]$data), method = "average")
}
pdf(file = "Plot_v3/SampleClustering.pdf", width = 12, height = 12);
par(mfrow=c(2,1))
par(mar = c(0, 4, 2, 0))
for (set in 1:nSets)
  plot(sampleTrees[[set]], main = paste("Sample clustering on all genes in", setLabels[set]),
    xlab="", sub="", cex = 0.7);
dev.off();
```

#Plot to see what soft-thresholding powers would be appropriate

```{r}
# Choose a set of soft-thresholding powers
powers = c(seq(4,10,by=1), seq(12,20, by=2))

# Initialize a list to hold the results of scale-free analysis
powerTables = vector(mode = "list", length = nSets)

# Call the network topology analysis function for each set in turn
#The aim of pickSoftThreshold function is to help the user pick an appropriate soft-thresholding power for network construction
for (set in 1:nSets)
powerTables[[set]] = list(data = pickSoftThreshold(multiExpr[[set]]$data, powerVector=powers, verbose = 2)[[2]])
collectGarbage()

# Plot the results:
colors = c("yellow", "red", "blue", "green","black")

# Will plot these columns of the returned scale free analysis tables
plotCols = c(2,5,6,7)
colNames = c("Scale Free Topology Model Fit", "Mean connectivity", "Median connectivity",
"Max connectivity")

# Get the minima and maxima of the plotted points
ylim = matrix(NA, nrow = 2, ncol = 4);
for (set in 1:nSet) {
  for (col in 1:length(plotCols)) {
    ylim[1, col] = min(ylim[1, col], powerTables[[set]]$data[, plotCols[col]], na.rm = TRUE);
    ylim[2, col] = max(ylim[2, col], powerTables[[set]]$data[, plotCols[col]], na.rm = TRUE);
  }
}
```

# Plot the quantities in the chosen columns vs. the soft thresholding power

```{r}
sizeGrWindow(8, 6)
pdf(file = "Plot_v3/scaleFreeAnalysis.pdf", wi = 8, he = 6)
par(mfcol = c(2,2));
par(mar = c(4.2, 4.2 , 2.2, 0.5))
cex1 = 0.7;
for (col in 1:length(plotCols)) for (set in 1:nSet) {
  if (set==1) {
    plot(powerTables[[set]]$data[,1], -sign(powerTables[[set]]$data[,3])*powerTables[[set]]$data[,2],
        xlab="Soft Threshold (power)",ylab=colNames[col],type="n", ylim = ylim[, col],
        main = colNames[col]);
    addGrid();
  }
  if (col==1) {
    text(powerTables[[set]]$data[,1], -sign(powerTables[[set]]$data[,3])*powerTables[[set]]$data[,2],
        labels=powers,cex=cex1,col=colors[set]);
  } else
  text(powerTables[[set]]$data[,1], powerTables[[set]]$data[,plotCols[col]],
      labels=powers,cex=cex1,col=colors[set]);

  if (col==1) {
    legend("bottomright", legend = setLabels, col = colors, pch = 20) ;
  } else
    legend("topright", legend = setLabels, col = colors, pch = 20) ;
}
dev.off();

```

#we will calculate all of the necessary values to run WGCNA
#choose 10

```{r}
softPower = 10
#all tissues
adjacency = adjacency(datExpr0tissues,power=softPower,type="signed"); #picked signed based on the WGCNA suggection. prefer signed vs unsigned
dim(adjacency) #36673 36673
diag(adjacency)=0
dissTOM   = 1-TOMsimilarity(adjacency, TOMType="signed")
dim(dissTOM) #36673 36673
save(adjacency, file = "adjacency.RData")
save(dissTOM, file = "dissTOM.Rdata")
#flower
adjacency.flower = adjacency(datExpr0flower,power=softPower,type="signed")
dim(adjacency.flower) #36673 36673
diag(adjacency.flower)=0
dissTOM.flower   = 1-TOMsimilarity(adjacency.flower, TOMType="signed")
save(adjacency.flower, file = "adjacency.flower.RData")
save(dissTOM.flower, file = "dissTOM.flower.Rdata")
#root
adjacency.root = adjacency(datExpr0root,power=softPower,type="signed")
dim(adjacency.root) #36673 36673
diag(adjacency.root)=0
dissTOM.root   = 1-TOMsimilarity(adjacency.root, TOMType="signed")
save(adjacency.root, file = "adjacency.root.RData")
save(dissTOM.root, file = "dissTOM.root.Rdata")
adjacency.stem = adjacency(datExpr0stem,power=softPower,type="signed")
dim(adjacency.stem) 
diag(adjacency.stem)=0
dissTOM.stem   = 1-TOMsimilarity(adjacency.stem, TOMType="signed")
save(adjacency.stem, file = "adjacency.stem.RData")
save(dissTOM.stem, file = "dissTOM.stem.Rdata")
```

#
```{r}
#all tissues
install.packages("flashClust")
library(flashClust)
geneTree  = flashClust(as.dist(dissTOM), method="average")
# display the networks visually
sizeGrWindow(6,16)
pdf("Plot_v3/dendrogram.pdf",height=6,width=16)
plot(geneTree,xlab="",sub="",main="Gene clustering on TOM-based dissimilarity", labels=FALSE,hang=0.04);
dev.off()  #These are "good" data, since there are a lot of distinct branches

#define modules, we will determine modules based on total data after removing batch effect
# We like large modules, so we set the minimum module size relatively high:
minModuleSize <- 30;
# Module identification using dynamic tree cut:
dynamicMods <- cutreeDynamic(dendro = geneTree, distM = dissTOM,
                             deepSplit <- 2, pamRespectsDendro = FALSE,
                            minClusterSize = minModuleSize);

table(dynamicMods)
table(dynamicMods) %>% length() # 20 modules 

# Convert numeric lables into colors
dynamicColors = labels2colors(dynamicMods) 
table(dynamicColors)
# Plot the dendrogram and colors CRderneath
sizeGrWindow(8,6)
pdf("Plot_v3/Final_modules.pdf",height=8,width=12)
plotDendroAndColors(geneTree, dynamicColors, "Dynamic Tree Cut",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "Gene dendrogram and module colors")
dev.off()
```

```{r}
#root
install.packages("flashClust")
library(flashClust)
geneTree.root  = flashClust(as.dist(dissTOM.root), method="average")
# display the networks visually
sizeGrWindow(6,16)
pdf("Plot_v3/dendrogram_root.pdf",height=6,width=16)
plot(geneTree.root,xlab="",sub="",main="Gene clustering on TOM-based dissimilarity", labels=FALSE,hang=0.04);
dev.off()  #These are "good" data, since there are a lot of distinct branches

#define modules, we will determine modules based on total data after removing batch effect
# We like large modules, so we set the minimum module size relatively high:
minModuleSize <- 30;
# Module identification using dynamic tree cut:
dynamicMods.root <- cutreeDynamic(dendro = geneTree.root, distM = dissTOM.root,
                             deepSplit <- 2, pamRespectsDendro = FALSE,
                            minClusterSize = minModuleSize);

table(dynamicMods.root)
table(dynamicMods.root) %>% length() # 37 modules 

# Convert numeric lables into colors
dynamicColors.root = labels2colors(dynamicMods.root) # see this for the num modules of each genes
table(dynamicColors.root)
# Plot the dendrogram and colors CRderneath
sizeGrWindow(8,6)
pdf("Plot_v3/Final_modules_root.pdf",height=8,width=12)
plotDendroAndColors(geneTree.root, dynamicColors.root, "Dynamic Tree Cut",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
               main = "Gene dendrogram and module colors")
dev.off()
```

```{r}
#flower
#install.packages("flashClust")
#library(flashClust)
geneTree.flower  = flashClust(as.dist(dissTOM.flower), method="average")
# display the networks visually
sizeGrWindow(6,16)
pdf("Plot_v3/dendrogram_flower.pdf",height=6,width=16)
plot(geneTree.flower,xlab="",sub="",main="Gene clustering on TOM-based dissimilarity", labels=FALSE,hang=0.04);
dev.off()  #These are "good" data, since there are a lot of distinct branches

#define modules, we will determine modules based on total data after removing batch effect
# We like large modules, so we set the minimum module size relatively high:
minModuleSize <- 30;
# Module identification using dynamic tree cut:
dynamicMods.flower <- cutreeDynamic(dendro = geneTree.flower, distM = dissTOM.flower,
                             deepSplit <- 2, pamRespectsDendro = FALSE,
                            minClusterSize = minModuleSize);

table(dynamicMods.flower)
table(dynamicMods.flower) %>% length() # 32 modules 

# Convert numeric lables into colors
dynamicColors.flower = labels2colors(dynamicMods.flower) # see this for the num modules of each genes
table(dynamicColors.flower)
# Plot the dendrogram and colors CRderneath
sizeGrWindow(8,6)
pdf("Plot_v3/Final_modules_flower.pdf",height=8,width=12)
plotDendroAndColors(geneTree.flower, dynamicColors.flower, "Dynamic Tree Cut",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
               main = "Gene dendrogram and module colors")
dev.off()
```

```{r}
#stem
#install.packages("flashClust")
#library(flashClust)
geneTree.stem  = flashClust(as.dist(dissTOM.stem), method="average")
# display the networks visually
sizeGrWindow(6,16)
pdf("Plot_v3/dendrogram_stem.pdf",height=6,width=16)
plot(geneTree.stem,xlab="",sub="",main="Gene clustering on TOM-based dissimilarity", labels=FALSE,hang=0.04);
dev.off()  #These are "good" data, since there are a lot of distinct branches

#define modules, we will determine modules based on total data after removing batch effect
# We like large modules, so we set the minimum module size relatively high:
minModuleSize <- 30;
# Module identification using dynamic tree cut:
dynamicMods.root <- cutreeDynamic(dendro = geneTree.root, distM = dissTOM.root,
                             deepSplit <- 2, pamRespectsDendro = FALSE,
                            minClusterSize = minModuleSize);

table(dynamicMods.root)
table(dynamicMods.root) %>% length() # 37 modules 

# Convert numeric lables into colors
dynamicColors.root = labels2colors(dynamicMods.root) # see this for the num modules of each genes
table(dynamicColors.root)
# Plot the dendrogram and colors CRderneath
sizeGrWindow(8,6)
pdf("Plot_v3/Final_modules_root.pdf",height=8,width=12)
plotDendroAndColors(geneTree.root, dynamicColors.root, "Dynamic Tree Cut",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
               main = "Gene dendrogram and module colors")
dev.off()
```

# merge similar modules
```{r}
#all tissues
# Calculate eigengenes 
MEList <- moduleEigengenes(datExpr0tissues, colors = dynamicColors) # Calculates module eigengenes (1st principal component) of modules in a given single dataset. 
MEs <- MEList$eigengenes
# Calculate dissimilarity of module eigengenes
MEDiss <- 1-cor(MEs);
# Cluster module eigengenes
METree <- hclust(as.dist(MEDiss), method = "average");
# Plot the result
sizeGrWindow(7, 6)
plot(METree, main = "Clustering of module eigengenes",
     xlab = "", sub = "")
# merge with correlation > 0.8
MEDissThres = 0.2
# Plot the cut line into the dendrogram
plot(METree, main = "Clustering of module eigengenes",
     xlab = "", sub = "")
abline(h=MEDissThres, col = "red")
# Call an automatic merging 
merge = mergeCloseModules(datExpr0tissues, dynamicColors, cutHeight = MEDissThres, verbose = 3)
# The merged module colors
mergedColors = merge$colors
# Eigengenes of the new merged modules:
mergedMEs = merge$newMEs 
# compare pre and post merge
sizeGrWindow(12, 9)
pdf(file = "Plot_v3/geneDendro.pdf", wi = 9, he = 6)
plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors),
c("Dynamic Tree Cut", "Merged dynamic"),
dendroLabels = FALSE, hang = 0.03,
addGuide = TRUE, guideHang = 0.05)
dev.off() 
# Rename to moduleColors
moduleColors = mergedColors
# Construct numerical labels corresponding to the colors
colorOrder = c("grey", standardColors(50));
moduleLabels = match(moduleColors, colorOrder)-1; #16 modules after mergeing
MEs = mergedMEs
save(MEs, file = "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/WGCNA_out/MEs_all_tissues.Rdata")
```
# merge similar modules

```{r}
#root
# Calculate eigengenes 
MEList.root <- moduleEigengenes(datExpr0root, colors = dynamicColors.root) # Calculates module eigengenes (1st principal component) of modules in a given single dataset. 
MEs.root <- MEList.root$eigengenes
# Calculate dissimilarity of module eigengenes
MEDiss.root <- 1-cor(MEs.root);
# Cluster module eigengenes
METree.root <- hclust(as.dist(MEDiss.root), method = "average");
# Plot the result
sizeGrWindow(7, 6)
plot(METree.root, main = "Clustering of module eigengenes",
     xlab = "", sub = "")
# merge with correlation > 0.8
MEDissThres = 0.2
# Plot the cut line into the dendrogram
plot(METree.root, main = "Clustering of module eigengenes",
     xlab = "", sub = "")
abline(h=MEDissThres, col = "red")
# Call an automatic merging 
merge.root = mergeCloseModules(datExpr0root, dynamicColors.root, cutHeight = MEDissThres, verbose = 3)
# The merged module colors
mergedColors.root = merge.root$colors
# Eigengenes of the new merged modules:
mergedMEs.root = merge.root$newMEs 
# compare pre and post merge
sizeGrWindow(12, 9)
pdf(file = "Plot_v3/geneDendro_root.pdf", wi = 9, he = 6)
plotDendroAndColors(geneTree.root, cbind(dynamicColors.root, mergedColors.root),
c("Dynamic Tree Cut", "Merged dynamic"),
dendroLabels = FALSE, hang = 0.03,
addGuide = TRUE, guideHang = 0.05)
dev.off() 
# Rename to moduleColors
moduleColors.root = mergedColors.root
# Construct numerical labels corresponding to the colors
colorOrder = c("grey", standardColors(50));
moduleLabels.root = match(moduleColors.root, colorOrder)-1; #24 modules after mergeing
MEs.root = mergedMEs.root
save(MEs.root, file = "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/WGCNA_out/MEs_all_root.Rdata")
```
# merge similar modules

```{r}
#flower
# Calculate eigengenes 
MEList.flower <- moduleEigengenes(datExpr0flower, colors = dynamicColors.flower) # Calculates module eigengenes (1st principal component) of modules in a given single dataset. 
MEs.flower <- MEList.flower$eigengenes
# Calculate dissimilarity of module eigengenes
MEDiss.flower <- 1-cor(MEs.flower);
# Cluster module eigengenes
METree.flower <- hclust(as.dist(MEDiss.flower), method = "average");
# Plot the result
sizeGrWindow(7, 6)
plot(METree.flower, main = "Clustering of module eigengenes",
     xlab = "", sub = "")
# merge with correlation > 0.8
MEDissThres = 0.2
# Plot the cut line into the dendrogram
plot(METree.flower, main = "Clustering of module eigengenes",
     xlab = "", sub = "")
abline(h=MEDissThres, col = "red")
# Call an automatic merging 
merge.flower = mergeCloseModules(datExpr0flower, dynamicColors.flower, cutHeight = MEDissThres, verbose = 3)
# The merged module colors
mergedColors.flower = merge.flower$colors
# Eigengenes of the new merged modules:
mergedMEs.flower = merge.flower$newMEs 
# compare pre and post merge
sizeGrWindow(12, 9)
pdf(file = "Plot_v3/geneDendro_flower.pdf", wi = 9, he = 6)
plotDendroAndColors(geneTree.flower, cbind(dynamicColors.flower, mergedColors.flower),
c("Dynamic Tree Cut", "Merged dynamic"),
dendroLabels = FALSE, hang = 0.03,
addGuide = TRUE, guideHang = 0.05)
dev.off() 
# Rename to moduleColors
moduleColors.flower = mergedColors.flower
# Construct numerical labels corresponding to the colors
colorOrder = c("grey", standardColors(50));
moduleLabels.flower = match(moduleColors.flower, colorOrder)-1; #21 modules after mergeing
MEs.flower = mergedMEs.flower
save(MEs.flower, file = "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/hajaramini/Ferula_RNAseq_Analysis/WGCNA_out/MEs_all_flower.Rdata")
```

#Next we calculate the principle components for visualizations(Julin's suggestion), see the mean of expression of flower, stem, root and leaf in tissue modules 

```{r}
#To see the expression pattern of all genes in all modules of tissue network (big net) in each root samples (#40 modules with 6 root samples)

dim(MEs) #12 45 
colors = names(table(moduleColors))
ordergenes = geneTree$order
pdf("Plot_v4/ModuleEigengeneVisualization.pdf",height=6,width=6)

par(mfrow=c(1,1), mar=c(6, 5, 2, 2) + 0.1, cex=1)

for (which.module in names(table(moduleColors))){
  ME = MEs[, paste("ME",which.module, sep="")] 
  barplot(ME, col=which.module, main="", cex.main=2, 
      ylab="eigengene expression",xlab="array sample") 
}; 

dev.off()
#we could not see the consistant across the tissue. seems still batch effect has strong role.
```
